### Peak calling with merged ATAC-seq samples using HMMRATAC

We have processed and QCed the ATAC-seq samples. We then used MACS2 to call peaks and used that output for all downstream analyses.

Now there is an ATAC-specific peak caller called HMMRATAC, developed by the author of MACS2, but exploiting the specific features of ATAC-seq data to improve peak calling.
https://academic.oup.com/nar/advance-article/doi/10.1093/nar/gkz533/5519166

We will use this new program to call peaks in all the merged samples that passed QC (50 samples).
For this we need one file with all the merged data, sorted and indexed. 

```{bash}
sh 05.1_HMMRATAC_prepare_mergedSamples.sh
```

And we also need a file with chromosome sizes:

```{bash}
samtools view -H /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/ATAC-seq/HMMRATAC/mergedSamples.bam | perl -ne 'if(/^@SQ.*?SN:(\w+)\s+LN:(\d+)/){print $1,"\t",$2,"\n"}' > /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/ATAC-seq/HMMRATAC/genome.info
```

Now we can run it:

```{bash}
sh 05.2_HMMRATAC_callPeaks_mergedSamples.sh
```

