---
title: "Differential expression analysis of mouse somites - comparison of methods"
date: "29 May 2019"
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(edgeR)
library(limma)
library(ggplot2)
library(ggpubr)
library(RColorBrewer)
library(VennDiagram)

dir <- "/user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/"

palette(c(brewer.pal(n=8, "Set2"), brewer.pal(n=9, "Set1")))

#### FUNCTIONS
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)
  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)
  numPlots = length(plots)
  
  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                     ncol = cols, nrow = ceiling(numPlots/cols), byrow = TRUE)
  }
  
  if (numPlots==1) {
    print(plots[[1]])
    
  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
    
    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
      
      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}
```

We have processed and QCed the RNA-seq from mouse somites. We then identified strong technical effects in the data, and thus used two different approaches to identify and remove such variation: surrogate variable analysis (SVA) and a PCA-based approach using the residuals of the fit of interest (combination of the somite and stage of each sample) to capture effects unrelated to this.

```{r data}
meta <- read.table(paste0(dir, "RNA-seq/data/metadata_RNAseq.tsv"), stringsAsFactors = FALSE, header = TRUE)
meta <- meta[meta$use==1,]

data <- read.table(paste0(dir, "RNA-seq/data/geneCounts.RAW.tsv"))
y <- readRDS(paste0(dir, "RNA-seq/results/01_edgeRobject.Rds"))
```

We then performed differential expression analysis in three modes:

- **Baseline**: no control for any technical effects; this will probably have a lot of false positives and miss many true positives.
- **SVA approach**: control technical variation by blocking on the identified SVs.
- **PCA approach**: control technical variation by blocking on the identified PCs.

So let's now compare the results from each.

#### Somite trios

First, we compare the somite I vs II vs III, which should identify the genes involved in triggering the differentiation program of somites after they are segmented.

We performed the test both on a stage-specific manner and taking the average across stages. To keep the comparisons simple, we will only focus on the averaged tests.

In all cases, we observed the most changes when comparing somites I and III, which are the most distinct. The number of genes detected was greater when using SVA compared to the baseline, and much higher with the PCA approach.

```{r somiteTrios}
## baseline
somiteTrios_base <- list()
somiteTrios_base[['IvsII']] <- read.table(paste0(dir, "RNA-seq/results/03.1_DEresults_somiteTrios_somiteIvsII.tsv"), stringsAsFactors = FALSE)
somiteTrios_base[['IIvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.1_DEresults_somiteTrios_somiteIIvsIII.tsv"), stringsAsFactors = FALSE)
somiteTrios_base[['IvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.1_DEresults_somiteTrios_somiteIvsIII.tsv"), stringsAsFactors = FALSE)

somiteTrios_base[["IvsII"]] <- somiteTrios_base[["IvsII"]][somiteTrios_base[["IvsII"]]$FDR < 0.05,]
somiteTrios_base[["IIvsIII"]] <- somiteTrios_base[["IIvsIII"]][somiteTrios_base[["IIvsIII"]]$FDR < 0.05,]
somiteTrios_base[["IvsIII"]] <- somiteTrios_base[["IvsIII"]][somiteTrios_base[["IvsIII"]]$FDR < 0.05,]

## SVA
somiteTrios_sva <- list()
somiteTrios_sva[['IvsII']] <- read.table(paste0(dir, "RNA-seq/results/03.2_DEresults_somiteTrios_somiteIvsII_sva.tsv"), stringsAsFactors = FALSE)
somiteTrios_sva[['IIvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.2_DEresults_somiteTrios_somiteIIvsIII_sva.tsv"), stringsAsFactors = FALSE)
somiteTrios_sva[['IvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.2_DEresults_somiteTrios_somiteIvsIII_sva.tsv"), stringsAsFactors = FALSE)

somiteTrios_sva[["IvsII"]] <- somiteTrios_sva[["IvsII"]][somiteTrios_sva[["IvsII"]]$FDR < 0.05,]
somiteTrios_sva[["IIvsIII"]] <- somiteTrios_sva[["IIvsIII"]][somiteTrios_sva[["IIvsIII"]]$FDR < 0.05,]
somiteTrios_sva[["IvsIII"]] <- somiteTrios_sva[["IvsIII"]][somiteTrios_sva[["IvsIII"]]$FDR < 0.05,]

## PCA
somiteTrios_pca <- list()
somiteTrios_pca[['IvsII']] <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_somiteTrios_somiteIvsII_pca.tsv"), stringsAsFactors = FALSE)
somiteTrios_pca[['IIvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_somiteTrios_somiteIIvsIII_pca.tsv"), stringsAsFactors = FALSE)
somiteTrios_pca[['IvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_somiteTrios_somiteIvsIII_pca.tsv"), stringsAsFactors = FALSE)

somiteTrios_pca[["IvsII"]] <- somiteTrios_pca[["IvsII"]][somiteTrios_pca[["IvsII"]]$FDR < 0.05,]
somiteTrios_pca[["IIvsIII"]] <- somiteTrios_pca[["IIvsIII"]][somiteTrios_pca[["IIvsIII"]]$FDR < 0.05,]
somiteTrios_pca[["IvsIII"]] <- somiteTrios_pca[["IvsIII"]][somiteTrios_pca[["IvsIII"]]$FDR < 0.05,]

nDE <- data.frame(unlist(lapply(somiteTrios_base, nrow)), unlist(lapply(somiteTrios_sva, nrow)), unlist(lapply(somiteTrios_pca, nrow)))
colnames(nDE) <- c("baseline", "SVA", "PCA")
nDE
```

Considering the somite I vs III comparison, the majority of the genes called with the base and SVA methods are a subset of the PCA approach. Overall, the results agree well.

```{r somiteTrios_overlap}
v <- venn.diagram(x=list(base=somiteTrios_base[['IvsIII']]$genes, SVA=somiteTrios_sva[['IvsIII']]$genes, PCA=somiteTrios_pca[['IvsIII']]$genes), fill=c("tomato1", "steelblue", "darkolivegreen2"), cat.col=c("tomato1", "steelblue", "darkolivegreen4"), alpha = 0.3, cat.fontface=2, filename = NULL)
grid.draw(v)
```

The genes identified only with the SVA method tend to have much higher FDRs. For the PCA-specific genes, there is a similar distribution of FDR values, with many more showing very small values. The comparison is more difficult given how large is the number of genes that are PCA-specific.

```{r somiteTrios_nonShared}
plots <- list()

shared <- somiteTrios_sva[["IvsIII"]][somiteTrios_sva[["IvsIII"]]$genes %in% intersect(somiteTrios_sva[["IvsIII"]]$genes, somiteTrios_pca[["IvsIII"]]$genes),]
sva_only <- somiteTrios_sva[["IvsIII"]][somiteTrios_sva[["IvsIII"]]$genes %in% setdiff(somiteTrios_sva[["IvsIII"]]$genes, somiteTrios_pca[["IvsIII"]]$genes),]

df <- data.frame(FDR = c(shared$FDR, sva_only$FDR), sva_genes=c(rep("shared", nrow(shared)), rep("unique", nrow(sva_only))))
plots[[1]] <- ggplot(df, aes(sva_genes, FDR)) + geom_violin(scale="width") + geom_boxplot(width=0.075)


shared <- somiteTrios_pca[["IvsIII"]][somiteTrios_pca[["IvsIII"]]$genes %in% intersect(somiteTrios_sva[["IvsIII"]]$genes, somiteTrios_pca[["IvsIII"]]$genes),]
pca_only <- somiteTrios_pca[["IvsIII"]][somiteTrios_pca[["IvsIII"]]$genes %in% setdiff(somiteTrios_pca[["IvsIII"]]$genes, somiteTrios_sva[["IvsIII"]]$genes),]

df <- data.frame(FDR = c(shared$FDR, pca_only$FDR), pca_genes=c(rep("shared", nrow(shared)), rep("unique", nrow(pca_only))))
plots[[2]] <- ggplot(df, aes(pca_genes, FDR)) + geom_violin(scale="width") + geom_boxplot(width=0.075)

multiplot(plotlist = plots, cols=2)
```

Thus, this suggests that the genes lost from the SVA method are weak canidadates, so not a bit loss.

The question now is whether the over 3 thousand genes recovered by the PCA approach are indeed DE. 

From the MA plot we can see that many of the fold-changes are very small.

```{r}
tmp <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_somiteTrios_somiteIvsIII_pca.tsv"), stringsAsFactors = FALSE)
colnames(tmp)[c(2,3,6)] <- c("log2FoldChange","baseMean","padj")
tmp$baseMean <- 2^tmp$baseMean-1
ggmaplot(tmp, FDR = 0.05, fc=0, genenames = tmp$genes, size=1, legend = "top", top=30, label.rectangle = TRUE, select.top.method = "padj", font.label = c("bold", 9), main="somite III vs I ", ggtheme = theme_minimal() )
```

Indeed, a large portion of the genes show a log2-fold-change of only 0.2 (dashed lines). This corresponds to a fold-change of `r round(2^0.2,2)` which is not really meaningful. This suggests that by removing all the structure found in the residuals of the fit we have reduced the noise to a point where we have ridiculous power to *detect tiny differences*, but these probably are not differences we are interested in. So we can add a filter requiring a fold-change of at least 1.5 (dotted lines) to retain only substantial changes.

```{r fc}
ggplot(somiteTrios_pca[["IvsIII"]], aes(1,logFC)) + geom_violin(lwd=0.75) + geom_hline(yintercept=c(-log2(1.5),-0.2,0.2,log2(1.5)), lty=c(3,2,2,3), col="grey")
```

This filter affects much more the results from the PCA analysis, followed by the SVA results and filters the fewest genes from the baseline results. This agrees with increases in power gained by reducing the dataset's varaibility.

```{r filter}
somiteTrios_base.filter <- somiteTrios_base
somiteTrios_base.filter <- lapply(somiteTrios_base.filter, function(x) x[abs(x$logFC) > log2(1.5),])

somiteTrios_sva.filter <- somiteTrios_sva
somiteTrios_sva.filter <- lapply(somiteTrios_sva.filter, function(x) x[abs(x$logFC) > log2(1.5),])

somiteTrios_pca.filter <- somiteTrios_pca
somiteTrios_pca.filter <- lapply(somiteTrios_pca.filter, function(x) x[abs(x$logFC) > log2(1.5),])

nDE <- cbind(nDE, data.frame(unlist(lapply(somiteTrios_base.filter, nrow)), unlist(lapply(somiteTrios_sva.filter, nrow)), unlist(lapply(somiteTrios_pca.filter, nrow))))
colnames(nDE)[-c(1:3)] <- paste0(colnames(nDE)[1:3], ">1.5")
nDE
```

After applying the minimum fold-change requirement, there are still many more genes identified with the PCA method. And now the results form the three analyses agree even better, with the vast majority of genes from the baseline and SVA tests identified in the PCA analysis also.

```{r somiteTrios_filter_overlap}
v <- venn.diagram(x=list(base=somiteTrios_base.filter[['IvsIII']]$genes, SVA=somiteTrios_sva.filter[['IvsIII']]$genes, PCA=somiteTrios_pca.filter[['IvsIII']]$genes), fill=c("tomato1", "steelblue", "darkolivegreen2"), cat.col=c("tomato1", "steelblue", "darkolivegreen4"), alpha = 0.3, cat.fontface=2, filename = NULL)
grid.draw(v)
```

So let's check the genes gained from the PCA method. As before, they have a similar distribution of p-values to those shared with the other methods, but they do tend to have smaller fold-changes, suggesting we gain them due to the increased power in the test.

```{r pca_only, fig.width=10}
plots <- list()
shared <- somiteTrios_pca.filter[["IvsIII"]][somiteTrios_pca.filter[["IvsIII"]]$genes %in% intersect(somiteTrios_sva.filter[["IvsIII"]]$genes, somiteTrios_pca.filter[["IvsIII"]]$genes),]
pca_only <- somiteTrios_pca.filter[["IvsIII"]][somiteTrios_pca.filter[["IvsIII"]]$genes %in% setdiff(somiteTrios_pca.filter[["IvsIII"]]$genes, somiteTrios_sva.filter[["IvsIII"]]$genes),]

df <- data.frame(FDR = c(shared$FDR, pca_only$FDR), pca_genes=c(rep("shared", nrow(shared)), rep("unique", nrow(pca_only))))
plots[[1]] <- ggplot(df, aes(pca_genes, FDR)) + geom_violin(scale="width") + geom_boxplot(width=0.075)

df <- data.frame(foldChange = c(shared$logFC, pca_only$logFC), pca_genes=c(rep("shared", nrow(shared)), rep("unique", nrow(pca_only))))
plots[[2]] <- ggplot(df, aes(pca_genes, foldChange)) + geom_violin(scale="width") + geom_boxplot(width=0.075)

multiplot(plotlist = plots, cols=2)
```

To convince ourselves that we are picking up sensible changes, we take genes from all levels of significance, that have only been identified with the PCA method.

```{r genes}
plot(pca_only[c(1:3,101:103,201:203,301:303,401:403,501:503,601:603,701:703),]$FDR, xlab="", ylab="FDR", axes=FALSE)
box(bty="l"); axis(2, las=2)
```

And we plot their expression levels in each somite, using baseline counts (that is, only normalised for depth of sequencing and composition biases, without removing the technical effects).

```{r retrieveExpr, fig.width=10, fig.height=20}
dataNorm <- cpm(y$counts, log=TRUE, prior.count = 1)
dataNorm <- cbind(data[match(row.names(dataNorm), row.names(data)),1], as.data.frame(dataNorm))
colnames(dataNorm)[1] <- "gene"
# dataNorm <- removeBatchEffect(dataNorm[,-1], design = model.matrix(~0+group, meta), covariates = pcs[,1:14])
# dataNorm <- cbind(data[match(row.names(dataNorm), row.names(data)),1], as.data.frame(dataNorm)); colnames(dataNorm)[1] <- "gene"

retrieveGeneExpr <- function(dataNorm, meta, columns=c("stage", "somite", "date"), gene){
  d <- t(as.data.frame(dataNorm[dataNorm$gene==gene,-1]))
  stopifnot(identical(row.names(d), meta$sample))
  d <- cbind(d, meta[,columns])
  colnames(d)[1] <- "count"
  return(d)
}

plots <- list()
j=1
for(i in pca_only$genes[c(1:3,101:103,201:203,301:303,401:403,501:503,601:603,701:703)]){
  d <- retrieveGeneExpr(dataNorm, meta, gene=i)
  plots[[j]] <- ggplot(d, aes(x=somite, y=count)) + geom_boxplot(aes(fill=somite)) + scale_fill_brewer(palette = "Purples") + ggtitle(i) + theme(legend.position="none") + ylab("CPM")
  j <- j+1
}
multiplot(plotlist = plots, cols=3)
```

Even the genes that have p-values close to 0.05 (last three) show clear dynamics across somites, even if the changes are quite subtle. Such differences will become more pronounced if the technical variation is regressed out:

```{r retrieveExpr_corr, fig.width=10, fig.height=20}
pcs <- read.table(paste0(dir, "RNA-seq/results/02_pcs_residuals.tab"))
dataNorm.corr <- removeBatchEffect(dataNorm[,-1], design = model.matrix(~0+group, meta), covariates = pcs[,1:14])
dataNorm.corr <- cbind(data[match(row.names(dataNorm.corr), row.names(data)),1], as.data.frame(dataNorm.corr)); colnames(dataNorm.corr)[1] <- "gene"

plots <- list()
j=1
for(i in pca_only$genes[c(1:3,101:103,201:203,301:303,401:403,501:503,601:603,701:703)]){
  d <- retrieveGeneExpr(dataNorm.corr, meta, gene=i)
  plots[[j]] <- ggplot(d, aes(x=somite, y=count)) + geom_boxplot(aes(fill=somite)) + scale_fill_brewer(palette = "Purples") + ggtitle(i) + theme(legend.position="none") + ylab("CPM")
  j <- j+1
}
multiplot(plotlist = plots, cols=3)
```

So this all seems sensible, and the gained genes do seem to be DE between somites.

#### Developmental changes

So now let's do the same for the tests looking at changes across developmental stages.

In this case we see many more significant genes in the baseline compared to the SVA approach. Nontheless, the PCA method still recovers twice as many genes compared to the baseline. This suggests maybe one of the SVs is capturing a developmental effect, instead of technical noise.

```{r stages}
stages_base <- read.table(paste0(dir, "RNA-seq/results/03.1_DEresults_stage_all.tsv"), stringsAsFactors = FALSE)
stages_sva <- read.table(paste0(dir, "RNA-seq/results/03.2_DEresults_stage_all_sva.tsv"), stringsAsFactors = FALSE)
stages_pca <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_stage_all_pca.tsv"), stringsAsFactors = FALSE)

stages_base <- stages_base[stages_base$FDR < 0.05,]
stages_sva <- stages_sva[stages_sva$FDR < 0.05,]
stages_pca <- stages_pca[stages_pca$FDR < 0.05,]

nDE <- data.frame(baseline=nrow(stages_base), SVA=nrow(stages_sva), PCA=nrow(stages_pca))
row.names(nDE) <- "stageChanges"
nDE
```

Like before, the results from the baseline and SVA tests are pretty much subsets of the PCA set. 

```{r stages_overlap}
v <- venn.diagram(x=list(base=stages_base$genes, SVA=stages_sva$genes, PCA=stages_pca$genes), fill=c("tomato1", "steelblue", "darkolivegreen2"), cat.col=c("tomato1", "steelblue", "darkolivegreen4"), alpha = 0.3, cat.fontface=2, filename = NULL)
grid.draw(v)
```

And the genes identified with only the baseline or SVA approach have quite high FDRs, while those unique to the PCA method have a similar distribution to the genes identified in all three tests.

```{r stages_nonShared, fig.width=10, fig.height=4}
plots <- list()

shared <- stages_base[stages_base$genes %in% intersect(stages_base$genes, stages_pca$genes),]
base_only <- stages_base[stages_base$genes %in% setdiff(stages_base$genes, stages_pca$genes),]

df <- data.frame(FDR = c(shared$FDR, base_only$FDR), base_genes=c(rep("shared", nrow(shared)), rep("unique", nrow(base_only))))
plots[[1]] <- ggplot(df, aes(base_genes, FDR)) + geom_violin(scale="width") + geom_boxplot(width=0.075)

shared <- stages_sva[stages_sva$genes %in% intersect(stages_sva$genes, stages_pca$genes),]
sva_only <- stages_sva[stages_sva$genes %in% setdiff(stages_sva$genes, stages_pca$genes),]

df <- data.frame(FDR = c(shared$FDR, sva_only$FDR), sva_genes=c(rep("shared", nrow(shared)), rep("unique", nrow(sva_only))))
plots[[2]] <- ggplot(df, aes(sva_genes, FDR)) + geom_violin(scale="width") + geom_boxplot(width=0.075)


shared <- stages_pca[stages_pca$genes %in% intersect(intersect(stages_sva$genes, stages_pca$genes), stages_base$genes),]
pca_only <- stages_pca[stages_pca$genes %in% setdiff(stages_pca$genes, union(stages_base$genes, stages_sva$genes)),]

df <- data.frame(FDR = c(shared$FDR, pca_only$FDR), pca_genes=c(rep("shared", nrow(shared)), rep("unique", nrow(pca_only))))
plots[[3]] <- ggplot(df, aes(pca_genes, FDR)) + geom_violin(scale="width") + geom_boxplot(width=0.075)

multiplot(plotlist = plots, cols=3)
```

As observed with the somite trio test, the fold-changes are dominated by very small changes, due to the increased power. The changes between stages are much larger, so now the bulk of the genes have fold-changes of about 1.3 (dashed line). This is still a very small difference and probably not very informative. In this case it might be good to filter a little more stringently, and retain only genes with a fold-change of 2 or larger (dotted line).

```{r stage_fc}
## get the largest fold-change from all pairwise comparisons
stages_pca$logFC.max <- sapply(1:nrow(stages_pca), function(x) stages_pca[x,which.max(abs(stages_pca[x,2:16]))+1])

ggplot(stages_pca, aes(1,logFC.max)) + geom_violin(lwd=0.75) + geom_hline(yintercept=c(-log2(2),-0.45,0.45,log2(2)), lty=c(3,2,2,3), col="grey")
```

This filter reduces the number of DE genes to about half, quite uniformly for all methods.

```{r stages_filter}
stages_base$logFC.max <- sapply(1:nrow(stages_base), function(x) stages_base[x,which.max(abs(stages_base[x,2:16]))+1])
stages_sva$logFC.max <- sapply(1:nrow(stages_sva), function(x) stages_sva[x,which.max(abs(stages_sva[x,2:16]))+1])

stages_base.filter <- stages_base[abs(stages_base$logFC.max) > log2(1.5),]
stages_sva.filter <- stages_sva[abs(stages_sva$logFC.max) > log2(1.5),]
stages_pca.filter <- stages_pca[abs(stages_pca$logFC.max) > log2(1.5),]

nDE <- rbind(nDE, data.frame(baseline=nrow(stages_base.filter), SVA=nrow(stages_sva.filter), PCA=nrow(stages_pca.filter)))
row.names(nDE)[2] <- "FC>2"
nDE
```

















```{r somiteTrios_filter_overlap}
v <- venn.diagram(x=list(base=somiteTrios_base.filter[['IvsIII']]$genes, SVA=somiteTrios_sva.filter[['IvsIII']]$genes, PCA=somiteTrios_pca.filter[['IvsIII']]$genes), fill=c("tomato1", "steelblue", "darkolivegreen2"), cat.col=c("tomato1", "steelblue", "darkolivegreen4"), alpha = 0.3, cat.fontface=2, filename = NULL)
grid.draw(v)
```

So let's check the genes gained from the PCA method. As before, they have a similar distribution of p-values to those shared with the other methods, but they do tend to have smaller fold-changes, suggesting we gain them due to the increased power in the test.

```{r pca_only, fig.width=10}
plots <- list()
shared <- somiteTrios_pca.filter[["IvsIII"]][somiteTrios_pca.filter[["IvsIII"]]$genes %in% intersect(somiteTrios_sva.filter[["IvsIII"]]$genes, somiteTrios_pca.filter[["IvsIII"]]$genes),]
pca_only <- somiteTrios_pca.filter[["IvsIII"]][somiteTrios_pca.filter[["IvsIII"]]$genes %in% setdiff(somiteTrios_pca.filter[["IvsIII"]]$genes, somiteTrios_sva.filter[["IvsIII"]]$genes),]

df <- data.frame(FDR = c(shared$FDR, pca_only$FDR), pca_genes=c(rep("shared", nrow(shared)), rep("unique", nrow(pca_only))))
plots[[1]] <- ggplot(df, aes(pca_genes, FDR)) + geom_violin(scale="width") + geom_boxplot(width=0.075)

df <- data.frame(foldChange = c(shared$logFC, pca_only$logFC), pca_genes=c(rep("shared", nrow(shared)), rep("unique", nrow(pca_only))))
plots[[2]] <- ggplot(df, aes(pca_genes, foldChange)) + geom_violin(scale="width") + geom_boxplot(width=0.075)

multiplot(plotlist = plots, cols=2)
```

To convince ourselves that we are picking up sensible changes, we take genes from all levels of significance, that have only been identified with the PCA method.

```{r genes}
plot(pca_only[c(1:3,101:103,201:203,301:303,401:403,501:503,601:603,701:703),]$FDR, xlab="", ylab="FDR", axes=FALSE)
box(bty="l"); axis(2, las=2)
```

And we plot their expression levels in each somite, using baseline counts (that is, only normalised for depth of sequencing and composition biases, without removing the technical effects).

```{r retrieveExpr, fig.width=10, fig.height=20}
dataNorm <- cpm(y$counts, log=TRUE, prior.count = 1)
dataNorm <- cbind(data[match(row.names(dataNorm), row.names(data)),1], as.data.frame(dataNorm))
colnames(dataNorm)[1] <- "gene"
# dataNorm <- removeBatchEffect(dataNorm[,-1], design = model.matrix(~0+group, meta), covariates = pcs[,1:14])
# dataNorm <- cbind(data[match(row.names(dataNorm), row.names(data)),1], as.data.frame(dataNorm)); colnames(dataNorm)[1] <- "gene"

retrieveGeneExpr <- function(dataNorm, meta, columns=c("stage", "somite", "date"), gene){
  d <- t(as.data.frame(dataNorm[dataNorm$gene==gene,-1]))
  stopifnot(identical(row.names(d), meta$sample))
  d <- cbind(d, meta[,columns])
  colnames(d)[1] <- "count"
  return(d)
}

plots <- list()
j=1
for(i in pca_only$genes[c(1:3,101:103,201:203,301:303,401:403,501:503,601:603,701:703)]){
  d <- retrieveGeneExpr(dataNorm, meta, gene=i)
  plots[[j]] <- ggplot(d, aes(x=somite, y=count)) + geom_boxplot(aes(fill=somite)) + scale_fill_brewer(palette = "Purples") + ggtitle(i) + theme(legend.position="none") + ylab("CPM")
  j <- j+1
}
multiplot(plotlist = plots, cols=3)
```

Even the genes that have p-values close to 0.05 (last three) show clear dynamics across somites, even if the changes are quite subtle. Such differences will become more pronounced if the technical variation is regressed out:

```{r retrieveExpr_corr, fig.width=10, fig.height=20}
pcs <- read.table(paste0(dir, "RNA-seq/results/02_pcs_residuals.tab"))
dataNorm.corr <- removeBatchEffect(dataNorm[,-1], design = model.matrix(~0+group, meta), covariates = pcs[,1:14])
dataNorm.corr <- cbind(data[match(row.names(dataNorm.corr), row.names(data)),1], as.data.frame(dataNorm.corr)); colnames(dataNorm.corr)[1] <- "gene"

plots <- list()
j=1
for(i in pca_only$genes[c(1:3,101:103,201:203,301:303,401:403,501:503,601:603,701:703)]){
  d <- retrieveGeneExpr(dataNorm.corr, meta, gene=i)
  plots[[j]] <- ggplot(d, aes(x=somite, y=count)) + geom_boxplot(aes(fill=somite)) + scale_fill_brewer(palette = "Purples") + ggtitle(i) + theme(legend.position="none") + ylab("CPM")
  j <- j+1
}
multiplot(plotlist = plots, cols=3)
```







```{r info}
sessionInfo()
```

