---
title: "Differential expression analysis of mouse somites - comparison of methods"
date: "29 May 2019"
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---



We have processed and QCed the RNA-seq from mouse somites. We then identified strong technical effects in the data, and thus used two different approaches to identify and remove such variation: surrogate variable analysis (SVA) and a PCA-based approach using the residuals of the fit of interest (combination of the somite and stage of each sample) to capture effects unrelated to this.


```r
meta <- read.table(paste0(dir, "RNA-seq/data/metadata_RNAseq.tsv"), stringsAsFactors = FALSE, header = TRUE)
meta <- meta[meta$use==1,]

data <- read.table(paste0(dir, "RNA-seq/data/geneCounts.RAW.tsv"))
y <- readRDS(paste0(dir, "RNA-seq/results/01_edgeRobject.Rds"))
```

We then performed differential expression analysis in three modes:

- **Baseline**: no control for any technical effects; this will probably have a lot of false positives and miss many true positives.
- **SVA approach**: control technical variation by blocking on the identified SVs.
- **PCA approach**: control technical variation by blocking on the identified PCs.

So let's now compare the results from each.

#### Somite trios

First, we compare the somite I vs II vs III, which should identify the genes involved in triggering the differentiation program of somites after they are segmented.

We performed the test both on a stage-specific manner and taking the average across stages. To keep the comparisons simple, we will only focus on the averaged tests.

In all cases, we observed the most changes when comparing somites I and III, which are the most distinct. The number of genes detected was greater when using SVA compared to the baseline, and much higher with the PCA approach.


```r
## baseline
somiteTrios_base <- list()
somiteTrios_base[['IvsII']] <- read.table(paste0(dir, "RNA-seq/results/03.1_DEresults_somiteTrios_somiteIvsII.tsv"), stringsAsFactors = FALSE)
somiteTrios_base[['IIvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.1_DEresults_somiteTrios_somiteIIvsIII.tsv"), stringsAsFactors = FALSE)
somiteTrios_base[['IvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.1_DEresults_somiteTrios_somiteIvsIII.tsv"), stringsAsFactors = FALSE)

somiteTrios_base[["IvsII"]] <- somiteTrios_base[["IvsII"]][somiteTrios_base[["IvsII"]]$FDR < 0.05,]
somiteTrios_base[["IIvsIII"]] <- somiteTrios_base[["IIvsIII"]][somiteTrios_base[["IIvsIII"]]$FDR < 0.05,]
somiteTrios_base[["IvsIII"]] <- somiteTrios_base[["IvsIII"]][somiteTrios_base[["IvsIII"]]$FDR < 0.05,]

## SVA
somiteTrios_sva <- list()
somiteTrios_sva[['IvsII']] <- read.table(paste0(dir, "RNA-seq/results/03.2_DEresults_somiteTrios_somiteIvsII_sva.tsv"), stringsAsFactors = FALSE)
somiteTrios_sva[['IIvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.2_DEresults_somiteTrios_somiteIIvsIII_sva.tsv"), stringsAsFactors = FALSE)
somiteTrios_sva[['IvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.2_DEresults_somiteTrios_somiteIvsIII_sva.tsv"), stringsAsFactors = FALSE)

somiteTrios_sva[["IvsII"]] <- somiteTrios_sva[["IvsII"]][somiteTrios_sva[["IvsII"]]$FDR < 0.05,]
somiteTrios_sva[["IIvsIII"]] <- somiteTrios_sva[["IIvsIII"]][somiteTrios_sva[["IIvsIII"]]$FDR < 0.05,]
somiteTrios_sva[["IvsIII"]] <- somiteTrios_sva[["IvsIII"]][somiteTrios_sva[["IvsIII"]]$FDR < 0.05,]

## PCA
somiteTrios_pca <- list()
somiteTrios_pca[['IvsII']] <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_somiteTrios_somiteIvsII_pca.tsv"), stringsAsFactors = FALSE)
somiteTrios_pca[['IIvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_somiteTrios_somiteIIvsIII_pca.tsv"), stringsAsFactors = FALSE)
somiteTrios_pca[['IvsIII']] <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_somiteTrios_somiteIvsIII_pca.tsv"), stringsAsFactors = FALSE)

somiteTrios_pca[["IvsII"]] <- somiteTrios_pca[["IvsII"]][somiteTrios_pca[["IvsII"]]$FDR < 0.05,]
somiteTrios_pca[["IIvsIII"]] <- somiteTrios_pca[["IIvsIII"]][somiteTrios_pca[["IIvsIII"]]$FDR < 0.05,]
somiteTrios_pca[["IvsIII"]] <- somiteTrios_pca[["IvsIII"]][somiteTrios_pca[["IvsIII"]]$FDR < 0.05,]

nDE <- data.frame(unlist(lapply(somiteTrios_base, nrow)), unlist(lapply(somiteTrios_sva, nrow)), unlist(lapply(somiteTrios_pca, nrow)))
colnames(nDE) <- c("baseline", "SVA", "PCA")
nDE
```

```
##         baseline SVA  PCA
## IvsII         95 408  731
## IIvsIII        3   0 2714
## IvsIII       700 860 4169
```

Considering the somite I vs III comparison, the majority of the genes called with the base and SVA methods are a subset of the PCA approach. Overall, the results agree well.


```r
v <- venn.diagram(x=list(base=somiteTrios_base[['IvsIII']]$genes, SVA=somiteTrios_sva[['IvsIII']]$genes, PCA=somiteTrios_pca[['IvsIII']]$genes), fill=c("tomato1", "steelblue", "darkolivegreen2"), alpha = 0.3, filename = NULL)
grid.draw(v)
```

![](03.4_differentialExpression_evaluationMethods_files/figure-html/somiteTrios_overlap-1.png)<!-- -->


```r
sva_only <- somiteTrios_sva[["IvsIII"]][somiteTrios_sva[["IvsIII"]]$genes %in% setdiff(somiteTrios_sva[["IvsIII"]]$genes, somiteTrios_pca[["IvsIII"]]$genes),]

boxplot(somiteTrios_sva[["IvsIII"]]$FDR, sva_only$FDR)
```

![](03.4_differentialExpression_evaluationMethods_files/figure-html/somiteTrios_svaOnly-1.png)<!-- -->
























```r
sessionInfo()
```

```
## R version 3.5.1 (2018-07-02)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS: /datastore/sw/gentoo2/usr/lib64/blas/reference/libblas.so.3.7.0
## LAPACK: /datastore/sw/gentoo2/usr/lib64/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] grid      stats     graphics  grDevices utils     datasets  methods  
## [8] base     
## 
## other attached packages:
## [1] VennDiagram_1.6.20  futile.logger_1.4.3 gplots_3.0.1.1     
## [4] RColorBrewer_1.1-2  ggplot2_3.1.1       edgeR_3.24.3       
## [7] limma_3.38.3       
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_1.0.1           formatR_1.6          pillar_1.3.1        
##  [4] compiler_3.5.1       plyr_1.8.4           futile.options_1.0.1
##  [7] bitops_1.0-6         tools_3.5.1          digest_0.6.18       
## [10] evaluate_0.13        tibble_2.1.1         gtable_0.3.0        
## [13] lattice_0.20-35      pkgconfig_2.0.2      rlang_0.3.4         
## [16] yaml_2.2.0           xfun_0.6             withr_2.1.2         
## [19] stringr_1.4.0        dplyr_0.8.0.1        knitr_1.22          
## [22] caTools_1.17.1.2     gtools_3.8.1         locfit_1.5-9.1      
## [25] tidyselect_0.2.5     glue_1.3.1           R6_2.4.0            
## [28] rmarkdown_1.12       gdata_2.18.0         lambda.r_1.2.3      
## [31] purrr_0.3.2          magrittr_1.5         scales_1.0.0        
## [34] htmltools_0.3.6      assertthat_0.2.1     colorspace_1.4-1    
## [37] KernSmooth_2.23-15   stringi_1.4.3        lazyeval_0.2.2      
## [40] munsell_0.5.0        crayon_1.3.4
```

