---
title: "Genes expressed dynamically during somitogenesis"
date: "30 May 2019"
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---



We have analysed the RNA-seq data from mouse somites, and defined a method based on PCA to capture the variation in the residuals of the fit of the biological question of interest, to remove technical variation. This allows identifying differentially expressed (DE) genes as somites differentiate and as development proceeds. So what are these genes and what processes are involved in somitogenesis?


```r
# metadata
meta <- read.table(paste0(dir, "RNA-seq/data/metadata_RNAseq.tsv"), stringsAsFactors = FALSE, header = TRUE)
meta <- meta[meta$use==1,]

# raw data
data <- read.table(paste0(dir, "RNA-seq/data/geneCounts.RAW.tsv"))
y <- readRDS(paste0(dir, "RNA-seq/results/01_edgeRobject.Rds"))

# normalised data
dataNorm <- cpm(y, log = TRUE, prior.count = 1)

# remove batch effect
pcs <- read.table(paste0(dir, "RNA-seq/results/02_pcs_residuals.tab"))
dataNorm <- removeBatchEffect(dataNorm, design = model.matrix(~0+group, meta), covariates = pcs[,1:14])
dataNorm <- cbind(data[match(row.names(dataNorm), row.names(data)),1], as.data.frame(dataNorm))
colnames(dataNorm)[1] <- "gene"
```

### Somite differentiation

After somites are segmented they begin a differentiation program to give rise to the appropriate derivatives. We have sequenced the three most recently segmented somites. These represent the somite that has just been segmented (less than 2 hours old) as well as somites that have been segmented for ~4 and 6 hours respectively; in these, the differentiation program should be underway.

So what can we learn about this process from the differentially expressed genes between the somite trios?

First, we filter the results from the DE analysis (PCA approach) to retain only genes with an adjusted p-value smaller than 0.05, and an absolute fold-change greater than 1.5. For cases where we tested several contrasts at once we retain the biggest fold-change and use that to filter.

We tested both stage-specific changes and the average across all stages. This resulted in a few hundred to around a thousand genes identified in each test.


```r
## stage-specific changes
somiteTrios.stage <- list()
for(contrast in paste0("stage", c(8,18,21,25,27,35))){
  somiteTrios.stage[[contrast]] <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_somiteTrios_", contrast, "_pca.tsv"), stringsAsFactors = FALSE)
  somiteTrios.stage[[contrast]] <- somiteTrios.stage[[contrast]][somiteTrios.stage[[contrast]]$FDR < 0.05,]
  somiteTrios.stage[[contrast]]$logFC.max <- sapply(1:nrow(somiteTrios.stage[[contrast]]), function(x) somiteTrios.stage[[contrast]][x,which.max(abs(somiteTrios.stage[[contrast]][x,2:4]))+1])
  somiteTrios.stage[[contrast]] <- somiteTrios.stage[[contrast]][abs(somiteTrios.stage[[contrast]]$logFC.max) > log2(1.5),]
}
nDE <- unlist(lapply(somiteTrios.stage, nrow))
nDE
```

```
##  stage8 stage18 stage21 stage25 stage27 stage35 
##     444     783     460     151     635     871
```

```r
## average across stages changes
somiteTrios.ave <- list()
for(contrast in paste0("somite", c("IvsII", "IIvsIII", "IvsIII"))){
  somiteTrios.ave[[contrast]] <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_somiteTrios_", contrast, "_pca.tsv"), stringsAsFactors = FALSE)
  somiteTrios.ave[[contrast]] <- somiteTrios.ave[[contrast]][somiteTrios.ave[[contrast]]$FDR < 0.05 & abs(somiteTrios.ave[[contrast]]$logFC) > log2(1.5),]
}
nDE <- unlist(lapply(somiteTrios.ave, nrow))
nDE
```

```
##   somiteIvsII somiteIIvsIII  somiteIvsIII 
##           294           442          1020
```

Many of the stage-specific changes are also identified when taking the average across stages. Nearly 1,300 genes are detected as DE in the average comparison, indicating changes that occur as somites differentiate, no matter how far along development this is happening. These represent the core network for somite differentiation/maturation.

However, there are also a number of genes that are identified only in stage-specific tests.


```r
somiteTrios.ave.all <- union(row.names(somiteTrios.ave[["somiteIvsII"]]), row.names(somiteTrios.ave[["somiteIIvsIII"]]))
somiteTrios.ave.all <- union(somiteTrios.ave.all, row.names(somiteTrios.ave[["somiteIvsIII"]]))
somiteTrios.all <- union(somiteTrios.ave.all, row.names(somiteTrios.stage[["stage8"]]))
somiteTrios.all <- union(somiteTrios.all, row.names(somiteTrios.stage[["stage18"]]))
somiteTrios.all <- union(somiteTrios.all, row.names(somiteTrios.stage[["stage21"]]))
somiteTrios.all <- union(somiteTrios.all, row.names(somiteTrios.stage[["stage25"]]))
somiteTrios.all <- union(somiteTrios.all, row.names(somiteTrios.stage[["stage27"]]))
somiteTrios.all <- union(somiteTrios.all, row.names(somiteTrios.stage[["stage35"]]))

somiteTrios.all <- as.data.frame(somiteTrios.all, stringsAsFactors=FALSE)
colnames(somiteTrios.all) <- "ID"
somiteTrios.all$gene <- data[somiteTrios.all$ID,1]
somiteTrios.all$average <- ifelse(somiteTrios.all$ID %in% somiteTrios.ave.all, 1, 0)
somiteTrios.all$stage8 <- ifelse(somiteTrios.all$ID %in% row.names(somiteTrios.stage[["stage8"]]) & somiteTrios.all$average == 0, 1, 0)
somiteTrios.all$stage18 <- ifelse(somiteTrios.all$ID %in% row.names(somiteTrios.stage[["stage18"]]) & somiteTrios.all$average == 0, 1, 0)
somiteTrios.all$stage21 <- ifelse(somiteTrios.all$ID %in% row.names(somiteTrios.stage[["stage21"]]) & somiteTrios.all$average == 0, 1, 0)
somiteTrios.all$stage25 <- ifelse(somiteTrios.all$ID %in% row.names(somiteTrios.stage[["stage25"]]) & somiteTrios.all$average == 0, 1, 0)
somiteTrios.all$stage27 <- ifelse(somiteTrios.all$ID %in% row.names(somiteTrios.stage[["stage27"]]) & somiteTrios.all$average == 0, 1, 0)
somiteTrios.all$stage35 <- ifelse(somiteTrios.all$ID %in% row.names(somiteTrios.stage[["stage35"]]) & somiteTrios.all$average == 0, 1, 0)

colSums(somiteTrios.all[,-c(1:2)])
```

```
## average  stage8 stage18 stage21 stage25 stage27 stage35 
##    1292     263     529     293      55     526     543
```

And ~70% of these are called in only one stage with the rest identified in two or more stages (but not on the average test). However, the vast majority of these genes are small differences that reach significance in only one or a few stages, but show consistent behaviour in most stages; thus, they represent cases of insufficient power. 

There are some cases of true stage specificity, but these are rare.


```r
upset(somiteTrios.all[somiteTrios.all$average==0,], nsets=6, nintersects = 100, sets.x.label = "number DE genes per stage")
```

![](03.5_differentiallyExpressedGenes_files/figure-html/overlap_stageSpecific-1.png)<!-- -->

#### Expression dynamics

From what we've observed, we expect most genes to be monotonically up/downregulated. To characterise the expression dynamics across the somite trios, we take the overall pattern across all stages, and fit a loess curve to each DE gene. we then cluster the curves to identify the prevalent patterns in the data.

A rough clustering results in 6 groups, which capture the general dynamics of expression.


```r
## fit loess
stopifnot(identical(meta$sample, colnames(dataNorm)[-1]))
somite <- as.factor(meta$somite)
dat <- 2^dataNorm[as.character(somiteTrios.all$ID),-1]
fits <- apply(dat, 1, function(x) stats::lowess(x~somite))

curves <- unique(as.data.frame(lapply(fits, '[', 2)))
colnames(curves) <- names(fits)
row.names(curves) <- paste0("somite",c("I","II","III"))

## cluster
curvesStd <- t(apply(t(curves), 1, function(x) x/max(x)))  # standarise
test <- cor(t(curvesStd), method="spearman")
test.dissim <- sqrt(0.5*((1-test)))
test.dist <- as.dist(test.dissim)
test.clust <- hclust(test.dist, method="average")

clust.genes <- cutreeDynamic(test.clust, distM=as.matrix(test.dist), minClusterSize=100, method="hybrid", deepSplit = 1, verbose = 0)
names(clust.genes) <- row.names(curvesStd)
# table(clust.genes)

plots <- list()
for(c in 1:max(clust.genes)){
  g <- names(clust.genes[clust.genes==c])
  test <- curvesStd[g,]
  mean <- colMeans(test)
  std <- apply(test, 2, sd)
  df.sub <- data.frame(x=1:3, ymin=mean-std, ymax=mean+std)
  df.sub.avg <- data.frame(x=1:3, y=mean)
  plots[[c]] <- ggplot() + geom_line(data=df.sub.avg, aes(x,y), colour="black" , size=1) +
    geom_ribbon(data=df.sub, aes(x=x,ymin=ymin,ymax=ymax), alpha=0.2, fill="black") +
    xlab(expression("somite")) + ylab(expression("expression")) + ggtitle(paste("Cluster", c, "-", nrow(test), "genes")) + ylim(0,1.2) +
    theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) + theme(legend.position = "none")
}
multiplot(plotlist = plots, cols = 3)
```

![](03.5_differentiallyExpressedGenes_files/figure-html/somiteTrios_fit-1.png)<!-- -->

A large portion of the expected genes monotonical up or downregulated along differentiation, but also many that are expressed differently in a particular somite.

Visualising every gene in each cluster reveals the dynamics of each more clearly:


```r
par(mfrow=c(2,3))
for(c in 1:max(clust.genes)){
  g <- names(clust.genes[clust.genes==c])
  test <- curvesStd[g,]
  plot(test[1,], type="l", axes=FALSE, xlab="", ylab="", main=paste("Cluster",c,"-",length(g),"genes"))
  box(bty="l"); axis(1, at=1:3, labels=c("SI","SII","SIII")); mtext(side=2, line=1, "relative expression", cex=0.8)
  for(i in 2:nrow(test)){
    par(new=T)
    plot(test[i,], type="l", axes=FALSE, xlab="", ylab="")
  }
}
```

![](03.5_differentiallyExpressedGenes_files/figure-html/somteTrios_clusters-1.png)<!-- -->

The genes being up or downregulated (first two clusters) do so with varying rate, where the increase/decay in expression is more or less abrupt. Similarly, we observe genes that are expressed the lowest in somiteI (cluster5), somiteII (cluster3 and 6) and somiteIII (cluster4) and that have different levels of expression in the other somites, from similarly high in both, to quite different levels. 


```r
heatmap.2(curvesStd[names(clust.genes[order(clust.genes)]),], trace="none", RowSideColors = as.character(clust.genes[names(clust.genes[order(clust.genes)])]), Rowv = FALSE, dendrogram = "none", col=brewer.pal(n=7, "Blues"), labRow = rep("", nrow(curvesStd)), cexCol = 0.8)
```

![](03.5_differentiallyExpressedGenes_files/figure-html/somiteTrios_heatmap-1.png)<!-- -->


#### Biological processes {.tabset}

Now let's explore what these DE genes are and their possible functions. We can use gene ontology enrichment analysis to determine this. We check for significant enrichments considering all 3171 genes that are significant in either the average or stage-specific tests. Not surprisingly, many terms related to development are highly significant. These include patterning and segmentation, differentiation and morphogenesis, and many different organs. 


```r
universe <- row.names(dataNorm)
all <- as.factor(as.numeric(universe %in% somiteTrios.all$ID))
names(all) <- universe

go.all <- new("topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl")
go.all.test <- runTest(go.all, algorithm = "classic", statistic = "Fisher" )
go.all.res <- GenTable(go.all, Fisher.classic = go.all.test, topNodes = length(score(go.all.test)))
go.all.res$Fisher.classic.adj <- p.adjust(go.all.res$Fisher.classic, "fdr")
go.all.res[c(3,5,9,10,13,14,19,20,22,23,24,33,34,38,39,40,46,49,53,59),]
```

```
##         GO.ID                                     Term Annotated
## 3  GO:0009653       anatomical structure morphogenesis      2271
## 5  GO:0048513                 animal organ development      2701
## 9  GO:0060429                   epithelium development       967
## 10 GO:0030154                     cell differentiation      3250
## 13 GO:0072359           circulatory system development       989
## 14 GO:0007155                            cell adhesion      1003
## 19 GO:0003002                          regionalization       343
## 20 GO:0035295                         tube development       974
## 22 GO:0007399               nervous system development      1922
## 23 GO:0007507                        heart development       558
## 24 GO:0007389            pattern specification process       429
## 33 GO:0001501              skeletal system development       457
## 34 GO:0050896                     response to stimulus      5864
## 38 GO:0009952 anterior/posterior pattern specification       218
## 39 GO:0060322                         head development       585
## 40 GO:0007417       central nervous system development       720
## 46 GO:0035282                             segmentation       104
## 49 GO:0001756                            somitogenesis        80
## 53 GO:0060537                muscle tissue development       404
## 59 GO:0001568                 blood vessel development       631
##    Significant Expected Fisher.classic Fisher.classic.adj
## 3          549   368.22        4.3e-27       1.081965e-23
## 5          629   437.94        8.9e-27       1.510508e-23
## 9          268   156.79        4.7e-21       4.431578e-18
## 10         703   526.95        1.0e-20       8.486000e-18
## 13         264   160.36        2.4e-18       1.566646e-15
## 14         265   162.63        8.8e-18       5.334057e-15
## 19         116    55.61        4.6e-16       2.054505e-13
## 20         253   157.92        5.1e-16       2.163930e-13
## 22         436   311.63        1.1e-15       4.243000e-13
## 23         164    90.47        1.3e-15       4.796435e-13
## 24         135    69.56        1.4e-15       4.950167e-13
## 33         137    74.10        6.0e-14       1.542909e-11
## 34        1112   950.79        7.7e-14       1.921829e-11
## 38          79    35.35        4.0e-13       8.932632e-11
## 39         161    94.85        1.1e-12       2.393487e-10
## 40         189   116.74        1.3e-12       2.757950e-10
## 46          47    16.86        3.1e-12       5.718826e-10
## 49          39    12.97        1.2e-11       2.078204e-09
## 53         118    65.50        2.3e-11       3.682604e-09
## 59         164   102.31        1.0e-10       1.414333e-08
```

```r
# using all <- as.factor(as.numeric(universe %in% somiteTrios.all[somiteTrios.all$average==1,]$ID)) instead, gives similar terms, but more more specialised terms and more related to somitogenes (although other organs are still there)
# using all <- as.factor(as.numeric(universe %in% somiteTrios.all[somiteTrios.all$average==0,]$ID)) instead, gives much more general terms, a lot of nervous-system terms and those related to somitogenesis are much rarer and less significant
```

Given the large redundance between GO terms, it'll take more detailed examination of the actual genes involved in these terms to determine their likely function in the context of the somites.

We can also restrict the analysis to the different clusters we've identified.

##### Cluster 1

These are genes that are downregulated as somites differentiate. The significant terms are very much related to somitogenesis, patterning and regionalisation, and less so to more general development terms. This suggests that many of the genes controlling the segmentation process are *lost* after the somite is segmented.


```r
all <- as.factor(as.numeric(universe %in% as.vector(names(clust.genes[clust.genes==1]))))
names(all) <- universe
# table(all)

go.clust1 <- new("topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl")
go.clust1.test <- runTest(go.clust1, algorithm = "classic", statistic = "Fisher" )
go.clust1.res <- GenTable(go.clust1, Fisher.classic = go.clust1.test, topNodes = length(score(go.clust1.test)))
go.clust1.res$Fisher.classic.adj <- p.adjust(go.clust1.res$Fisher.classic, "fdr")
go.clust1.res[c(1:6,9,11,12,16,17,20,21,24,27),]
```

```
##         GO.ID                                        Term Annotated
## 1  GO:0009952    anterior/posterior pattern specification       218
## 2  GO:0003002                             regionalization       343
## 3  GO:0007389               pattern specification process       429
## 4  GO:0009888                          tissue development      1580
## 5  GO:0060429                      epithelium development       967
## 6  GO:0009653          anatomical structure morphogenesis      2271
## 9  GO:0001756                               somitogenesis        80
## 11 GO:0035282                                segmentation       104
## 12 GO:0009887                  animal organ morphogenesis       890
## 16 GO:0061053                          somite development        99
## 17 GO:2000050 regulation of non-canonical Wnt signalin...        24
## 20 GO:0030154                        cell differentiation      3250
## 21 GO:0007399                  nervous system development      1922
## 24 GO:0001501                 skeletal system development       457
## 27 GO:0035108                          limb morphogenesis       164
##    Significant Expected Fisher.classic Fisher.classic.adj
## 1           37    10.97        6.2e-11       5.261320e-07
## 2           47    17.26        3.5e-10       1.485050e-06
## 3           51    21.59        9.1e-09       2.545800e-05
## 4          129    79.52        1.2e-08       2.545800e-05
## 5           88    48.67        3.1e-08       4.667300e-05
## 6          169   114.29        3.3e-08       4.667300e-05
## 9           17     4.03        3.6e-07       3.394400e-04
## 11          19     5.23        9.0e-07       6.943091e-04
## 12          76    44.79        3.5e-06       2.349969e-03
## 16          17     4.98        8.1e-06       4.296038e-03
## 17           8     1.21        1.4e-05       6.988471e-03
## 20         210   163.56        2.2e-05       9.334600e-03
## 21         135    96.73        2.5e-05       1.002891e-02
## 24          43    23.00        5.8e-05       2.002696e-02
## 27          21     8.25        7.6e-05       2.388652e-02
```

##### Cluster 2

This cluster contains the genes that are upregulated as maturation proceeds. In this case, the significant terms are much more related to adhesion and migration, broad developmental processes and organogenesis, and vasculature development. This suggests cells are acquiring the properties to reorganise themselves and migrate.


```r
all <- as.factor(as.numeric(universe %in% names(clust.genes[clust.genes==2])))
names(all) <- universe
# table(all)

go.clust2 <- new("topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl")
go.clust2.test <- runTest(go.clust2, algorithm = "classic", statistic = "Fisher" )
go.clust2.res <- GenTable(go.clust2, Fisher.classic = go.clust2.test, topNodes = length(score(go.clust2.test)))
go.clust2.res$Fisher.classic.adj <- p.adjust(go.clust2.res$Fisher.classic, "fdr")
go.clust2.res[c(1,5,10,11,12,15,16,18,20,24,29,39,47,48,54,70),]
```

```
##         GO.ID                               Term Annotated Significant
## 1  GO:0048856   anatomical structure development      4512         277
## 5  GO:0007155                      cell adhesion      1003          89
## 10 GO:0048513           animal organ development      2701         178
## 11 GO:0035295                   tube development       974          84
## 12 GO:0030154               cell differentiation      3250         204
## 15 GO:0072359     circulatory system development       989          82
## 16 GO:0040011                         locomotion      1428         104
## 18 GO:0030198  extracellular matrix organization       215          29
## 20 GO:0045595 regulation of cell differentiation      1566         110
## 24 GO:0001944            vasculature development       661          58
## 29 GO:0060429             epithelium development       967          75
## 39 GO:0050896               response to stimulus      5864         309
## 47 GO:0060322                   head development       585          49
## 48 GO:0001501        skeletal system development       457          41
## 54 GO:0007417 central nervous system development       720          56
## 70 GO:0043588                   skin development       205          22
##    Expected Fisher.classic Fisher.classic.adj
## 1    193.50        3.6e-13       3.054960e-09
## 5     43.01        2.7e-11       4.582440e-08
## 10   115.84        3.3e-10       2.800380e-07
## 11    41.77        4.3e-10       3.317255e-07
## 12   139.38        6.7e-10       4.738017e-07
## 15    42.41        4.6e-09       2.602373e-06
## 16    61.24        3.5e-08       1.856313e-05
## 18     9.22        4.2e-08       1.980067e-05
## 20    67.16        8.9e-08       3.776270e-05
## 24    28.35        1.6e-07       5.657333e-05
## 29    41.47        3.4e-07       9.900333e-05
## 39   251.48        9.9e-07       2.154138e-04
## 47    25.09        5.5e-06       9.930426e-04
## 48    19.60        6.4e-06       1.131467e-03
## 54    30.88        1.1e-05       1.728630e-03
## 70     8.79        7.1e-05       8.486000e-03
```

##### Cluster 4

These genes are expressed early (somites I and II) but are downregulated in the most mature somite. There are no terms that reach significant after multitesting correction.


```r
all <- as.factor(as.numeric(universe %in% names(clust.genes[clust.genes==4])))
names(all) <- universe
# table(all)

go.clust4 <- new("topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl")
go.clust4.test <- runTest(go.clust4, algorithm = "classic", statistic = "Fisher" )
go.clust4.res <- GenTable(go.clust4, Fisher.classic = go.clust4.test, topNodes = length(score(go.clust4.test)))
go.clust4.res$Fisher.classic.adj <- p.adjust(go.clust4.res$Fisher.classic, "fdr")
```

##### Cluster 5

These genes are expressed late (somites II and III) but are downregulated in the most immature somite. There are no terms that reach significant after multitesting correction.


```r
all <- as.factor(as.numeric(universe %in% names(clust.genes[clust.genes==5])))
names(all) <- universe
# table(all)

go.clust5 <- new("topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl")
go.clust5.test <- runTest(go.clust5, algorithm = "classic", statistic = "Fisher" )
go.clust5.res <- GenTable(go.clust5, Fisher.classic = go.clust5.test, topNodes = length(score(go.clust5.test)))
go.clust5.res$Fisher.classic.adj <- p.adjust(go.clust5.res$Fisher.classic, "fdr")
```

##### Clusters 3 and 6

These genes are expressed the lowest in the middle somite, with varying levels in the other two. The significant terms are a combination of those seen in clusters 1 and 2, including both segmentation and patterning related terms, and adhesion, migration and organogenesis terms. But also, some more specific terms related to development of the derivatives of somites.


```r
all <- as.factor(as.numeric(universe %in% names(clust.genes[clust.genes %in% c(3,6)])))
names(all) <- universe
# table(all)

go.clust6 <- new("topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl")
go.clust6.test <- runTest(go.clust6, algorithm = "classic", statistic = "Fisher" )
go.clust6.res <- GenTable(go.clust6, Fisher.classic = go.clust6.test, topNodes = length(score(go.clust6.test)))
go.clust6.res$Fisher.classic.adj <- p.adjust(go.clust6.res$Fisher.classic, "fdr")
go.clust6.res[c(3,6,7,10,11,13,14,15,17,19,28,41,45,64,69,70,78,82,97,100,103,105,107),]
```

```
##          GO.ID                                        Term Annotated
## 3   GO:0048856            anatomical structure development      4512
## 6   GO:0003002                             regionalization       343
## 7   GO:0045165                        cell fate commitment       226
## 10  GO:0007389               pattern specification process       429
## 11  GO:0030154                        cell differentiation      3250
## 13  GO:0009888                          tissue development      1580
## 14  GO:0048468                            cell development      1908
## 15  GO:0007155                               cell adhesion      1003
## 17  GO:0007417          central nervous system development       720
## 19  GO:0009952    anterior/posterior pattern specification       218
## 28  GO:0009887                  animal organ morphogenesis       890
## 41  GO:0055024 regulation of cardiac muscle tissue deve...        89
## 45  GO:0060537                   muscle tissue development       404
## 64  GO:0001756                               somitogenesis        80
## 69  GO:0061053                          somite development        99
## 70  GO:0060429                      epithelium development       967
## 78  GO:0035282                                segmentation       104
## 82  GO:0043588                            skin development       205
## 97  GO:0040008                        regulation of growth       621
## 100 GO:0042692                 muscle cell differentiation       333
## 103 GO:0014807                 regulation of somitogenesis        10
## 105 GO:0008544                       epidermis development       225
## 107 GO:0060538           skeletal muscle organ development       172
##     Significant Expected Fisher.classic Fisher.classic.adj
## 3           225   153.92        5.2e-12       1.470907e-08
## 6            38    11.70        1.5e-10       2.121500e-07
## 7            29     7.71        7.9e-10       9.577057e-07
## 10           41    14.63        2.4e-09       1.928636e-06
## 11          167   110.87        2.5e-09       1.928636e-06
## 13           97    53.90        4.4e-09       2.872185e-06
## 14          110    65.09        1.1e-08       6.667571e-06
## 15           69    34.21        1.5e-08       7.955625e-06
## 17           55    24.56        1.6e-08       7.986824e-06
## 19           26     7.44        2.8e-08       1.250568e-05
## 28           60    30.36        2.9e-07       8.789071e-05
## 41           14     3.04        1.8e-06       3.636857e-04
## 45           33    13.78        3.3e-06       6.223067e-04
## 64           12     2.73        1.6e-05       2.121500e-03
## 69           13     3.38        3.1e-05       3.812551e-03
## 70           57    32.99        3.4e-05       4.121771e-03
## 78           13     3.55        5.2e-05       5.657333e-03
## 82           19     6.99        7.7e-05       7.968561e-03
## 97           39    21.18        0.00018       1.558653e-02
## 100          25    11.36        0.00020       1.680396e-02
## 103           4     0.34        0.00024       1.977320e-02
## 105          19     7.68        0.00026       2.101295e-02
## 107          16     5.87        0.00027       2.141327e-02
```

For these genes, it is unclear how significant is the downregulation in somiteII and whether this is necessary or instead a reflection of highly variable, maybe noisier, genes.


```r
saveRDS(somiteTrios.stage, paste0(dir, "RNA-seq/results/03.5_DEgenes_somiteTrios_perStage_pca_FDR0.05_logFC1.5.Rds"))
saveRDS(somiteTrios.ave, paste0(dir, "RNA-seq/results/03.5_DEgenes_somiteTrios_average_pca_FDR0.05_logFC1.5.Rds"))
write.table(somiteTrios.all, paste0(dir, "RNA-seq/results/03.5_DEgenes_somiteTrios_all.tsv"), quote=FALSE, sep="\t", row.names = FALSE)
write.table(clust.genes, paste0(dir, "RNA-seq/results/03.5_DEgenes_somiteTrios_all_clusters.tsv"), quote=FALSE, sep="\t", col.names=FALSE, row.names=FALSE)
write.table(go.all.res, paste0(dir, "RNA-seq/results/03.5_DEgenes_somiteTrios_all_GOresults.tsv"), quote = FALSE, sep="\t", row.names = FALSE)
```

### Developmental changes

After somites are segmented they start differentiating and mature to eventually produced all the necessary derivatives that will create bone, muscle and skin, among other structures. Together with this process, the embryo is growing and developing, so there is a concomitant progression in developmental age, and the somites generated at different stages of development will generate slightly different structures.

To study this, we now use differential expression analysis to identify the genes that are dynamic across developmental time. We test this both in a somite-specific manner, and taking the average of all three somites from the same stage and recovering genes that are different between any pair of stages.


```r
## somite-specific changes
stages.somite <- list()
for(contrast in paste0("somite", c("I", "II", "III"))){
  stages.somite[[contrast]] <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_stage_",contrast,"_pca.tsv"), stringsAsFactors = FALSE)
  stages.somite[[contrast]] <- stages.somite[[contrast]][stages.somite[[contrast]]$FDR < 0.05,]
  stages.somite[[contrast]]$logFC.max <- sapply(1:nrow(stages.somite[[contrast]]), function(x) stages.somite[[contrast]][x,which.max(abs(stages.somite[[contrast]][x,2:16]))+1])
  stages.somite[[contrast]] <- stages.somite[[contrast]][abs(stages.somite[[contrast]]$logFC.max) > log2(1.5),]
}
nDE <- unlist(lapply(stages.somite, nrow))
nDE
```

```
##   somiteI  somiteII somiteIII 
##      5261      5415      7042
```

```r
## average across somites
stages.ave <- read.table(paste0(dir, "RNA-seq/results/03.3_DEresults_stage_all_pca.tsv"), stringsAsFactors = FALSE)
stages.ave <- stages.ave[stages.ave$FDR<0.05,]
stages.ave$logFC.max <- sapply(1:nrow(stages.ave), function(x) stages.ave[x,which.max(abs(stages.ave[x,2:16]))+1])
stages.ave <- stages.ave[abs(stages.ave$logFC.max) > log2(1.5),]
nDE <- nrow(stages.ave)
names(nDE) <- "average"
nDE
```

```
## average 
##    7890
```

Taking the somite trios as replicates is reasonable given that we've seen only a few thousand genes are DE between them, and their differences are quite small. Still, we first check whether there are changes that only occur in a given somite level.

While the majority of the genes significant in a specific somite level are also significant in the averaged test, there are around a thousand genes per somite that are not identified when taking the average.


```r
stages.all <- union(row.names(stages.ave), row.names(stages.somite[["somiteI"]]))
stages.all <- union(stages.all, row.names(stages.somite[["somiteII"]]))
stages.all <- union(stages.all, row.names(stages.somite[["somiteIII"]]))

stages.all <- as.data.frame(stages.all, stringsAsFactors=FALSE)
colnames(stages.all) <- "ID"
stages.all$gene <- data[as.character(stages.all$ID),1]
stages.all$average <- ifelse(stages.all$ID %in% row.names(stages.ave), 1, 0)
stages.all$somiteI <- ifelse(stages.all$ID %in% row.names(stages.somite[["somiteI"]]) & stages.all$average == 0, 1, 0)
stages.all$somiteII <- ifelse(stages.all$ID %in% row.names(stages.somite[["somiteII"]]) & stages.all$average == 0, 1, 0)
stages.all$somiteIII <- ifelse(stages.all$ID %in% row.names(stages.somite[["somiteIII"]]) & stages.all$average == 0, 1, 0)
colSums(stages.all[,-c(1:2)])
```

```
##   average   somiteI  somiteII somiteIII 
##      7890       895      1133      1791
```

And from these, the majority are identified in only one somite level.


```r
venn(data=list(somiteI=stages.all[stages.all$somiteI==1,1], somiteII=stages.all[stages.all$somiteII==1,1], somiteIII=stages.all[stages.all$somiteIII==1,1]))
```

![](03.5_differentiallyExpressedGenes_files/figure-html/somite-specific_overlap-1.png)<!-- -->

However, as before, more often than not, genes significant in only one somite show the same or similar trends in the other somites, just don't reach significance. Overall, it seems again that the somite-specific changes are very few.

#### Expression dynamics

Among the large number of genes changing expression across development, we have representation of many different patterns and thus becomes much more difficult to group them into congruent clusters. Nonetheless, a rough clustering shows that genes not only are up and downregulated as development proceeds, but also show higher expression at many different stages or even dynamic up-down-up regulation patterns.


```r
## fit loess
stopifnot(identical(meta$sample, colnames(dataNorm)[-1]))
stage <- factor(paste0("stage",meta$stage))
dat <- 2^dataNorm[as.character(stages.all$ID),-1]
fits <- apply(dat, 1, function(x) stats::lowess(x~stage))

curves <- unique(as.data.frame(lapply(fits, '[', 2)))
colnames(curves) <- names(fits)
row.names(curves) <- paste0("stage",c(8,18,21,25,27,35))

## cluster
curvesStd <- t(apply(t(curves), 1, function(x) x/max(x)))  # standarise
test <- cor(t(curvesStd), method="spearman")
test.dissim <- sqrt(0.5*((1-test)))
test.dist <- as.dist(test.dissim)
test.clust <- hclust(test.dist, method="average")

clust.genes <- cutreeDynamic(test.clust, distM=as.matrix(test.dist), minClusterSize=200, method="hybrid", deepSplit = 1, verbose = 0)
names(clust.genes) <- row.names(curvesStd)
# table(clust.genes)

plots <- list()
for(c in 1:max(clust.genes)){
  g <- names(clust.genes[clust.genes==c])
  test <- curvesStd[g,]
  mean <- colMeans(test)
  std <- apply(test, 2, sd)
  df.sub <- data.frame(x=1:6, ymin=mean-std, ymax=mean+std)
  df.sub.avg <- data.frame(x=1:6, y=mean)
  plots[[c]] <- ggplot() + geom_line(data=df.sub.avg, aes(x,y), colour="black" , size=1) +
    geom_ribbon(data=df.sub, aes(x=x,ymin=ymin,ymax=ymax), alpha=0.2, fill="black") +
    xlab(expression("somite")) + ylab(expression("expression")) + ggtitle(paste("Cluster", c, "-", nrow(test), "genes")) + ylim(0,1.2) +
    theme_bw() + theme(panel.border = element_blank(), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black")) + theme(legend.position = "none")
}
multiplot(plotlist = plots, cols = 4)
```

![](03.5_differentiallyExpressedGenes_files/figure-html/stages_fit-1.png)<!-- -->




#### Biological processes

Not surprisingly, DE genes across development are highly enriched for development-related terms. Terms are quite general and we will need to sub classify the genes into better defined patterns to find more meaningful enrichments.


```r
all <- as.factor(as.numeric(universe %in% stages.all$ID))
names(all) <- universe
# table(all)

go.all <- new("topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl")
go.all.test <- runTest(go.all, algorithm = "classic", statistic = "Fisher" )
go.all.res <- GenTable(go.all, Fisher.classic = go.all.test, topNodes = length(score(go.all.test)))
go.all.res$Fisher.classic.adj <- p.adjust(go.all.res$Fisher.classic, "fdr")
go.all.res[c(3,5,9,10,13,14,19,20,22,23,24,33,34,38,39,40,46,49,53,59),]
```

```
##         GO.ID                                        Term Annotated
## 3  GO:0032501            multicellular organismal process      5238
## 5  GO:0048856            anatomical structure development      4512
## 9  GO:0051239 regulation of multicellular organismal p...      2536
## 10 GO:0040011                                  locomotion      1428
## 13 GO:0009605               response to external stimulus      1531
## 14 GO:0023052                                   signaling      4109
## 19 GO:0048870                               cell motility      1268
## 20 GO:0051674                        localization of cell      1268
## 22 GO:0048869              cellular developmental process      3406
## 23 GO:0035295                            tube development       974
## 24 GO:0060429                      epithelium development       967
## 33 GO:0022603 regulation of anatomical structure morph...       925
## 34 GO:0001944                     vasculature development       661
## 38 GO:2000145                 regulation of cell motility       777
## 39 GO:0030334                regulation of cell migration       744
## 40 GO:0051240 positive regulation of multicellular org...      1492
## 46 GO:0048468                            cell development      1908
## 49 GO:0040012                    regulation of locomotion       836
## 53 GO:0051093 negative regulation of developmental pro...       873
## 59 GO:0030155                 regulation of cell adhesion       554
##    Significant Expected Fisher.classic Fisher.classic.adj
## 3         3248  2900.81        < 1e-30                 NA
## 5         2822  2498.75        < 1e-30                 NA
## 9         1654  1404.44        9.1e-29       3.857945e-25
## 10         983   790.83        4.9e-28       1.384903e-24
## 13        1041   847.87        1.2e-26       1.695800e-23
## 14        2560  2275.57        1.4e-26       1.695800e-23
## 19         877   702.22        5.6e-26       3.652492e-23
## 20         877   702.22        5.6e-26       3.652492e-23
## 22        2149  1886.25        1.0e-25       5.652667e-23
## 23         692   539.40        2.0e-25       1.059875e-22
## 24         687   535.53        3.2e-25       1.596047e-22
## 33         648   512.27        2.4e-21       7.826769e-19
## 34         481   366.06        3.3e-21       1.036322e-18
## 38         552   430.30        2.4e-20       6.564387e-18
## 39         531   412.03        2.7e-20       7.154156e-18
## 40         991   826.27        3.2e-20       8.222061e-18
## 46        1236  1056.65        2.2e-19       4.783026e-17
## 49         582   462.98        2.8e-18       5.652667e-16
## 53         599   483.47        1.2e-16       2.211913e-14
## 59         396   306.81        1.5e-15       2.445865e-13
```

For example, if we restrict the analysis to genes in cluster 10, which show a downregulation as development proceeds, terms are a bit more informative (although most are not significant).


```r
all <- as.factor(as.numeric(universe %in% as.vector(names(clust.genes[clust.genes==10]))))
names(all) <- universe
# table(all)

go.clust10 <- new("topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl")
go.clust10.test <- runTest(go.clust10, algorithm = "classic", statistic = "Fisher" )
go.clust10.res <- GenTable(go.clust10, Fisher.classic = go.clust10.test, topNodes = length(score(go.clust10.test)))
go.clust10.res$Fisher.classic.adj <- p.adjust(go.clust10.res$Fisher.classic, "fdr")
go.clust10.res[c(1,2,5,10,15,18),]
```

```
##         GO.ID                                        Term Annotated
## 1  GO:0035095             behavioral response to nicotine         7
## 2  GO:0060073                                 micturition        10
## 5  GO:0010810       regulation of cell-substrate adhesion       184
## 10 GO:0042273          ribosomal large subunit biogenesis        69
## 15 GO:0045198 establishment of epithelial cell apical/...        14
## 18 GO:0061158        3'-UTR-mediated mRNA destabilization        15
##    Significant Expected Fisher.classic Fisher.classic.adj
## 1            2     0.09         0.0030                  1
## 2            2     0.12         0.0063                  1
## 5            7     2.26         0.0077                  1
## 10           4     0.85         0.0102                  1
## 15           2     0.17         0.0124                  1
## 18           2     0.18         0.0142                  1
```


```r
saveRDS(stages.somite, paste0(dir, "RNA-seq/results/03.5_DEgenes_stages_perSomite_pca_FDR0.05_logFC1.5.Rds"))
saveRDS(stages.ave, paste0(dir, "RNA-seq/results/03.5_DEgenes_stages_average_pca_FDR0.05_logFC1.5.Rds"))
write.table(stages.all, paste0(dir, "RNA-seq/results/03.5_DEgenes_stages_all.tsv"), quote=FALSE, sep="\t", row.names = FALSE)
write.table(clust.genes, paste0(dir, "RNA-seq/results/03.5_DEgenes_stages_all_clusters.tsv"), quote=FALSE, sep="\t", col.names=FALSE, row.names=FALSE)
write.table(go.all.res, paste0(dir, "RNA-seq/results/03.5_DEgenes_stages_all_GOresults.tsv"), quote = FALSE, sep="\t", row.names = FALSE)
```




```r
sessionInfo()
```

```
## R version 3.5.1 (2018-07-02)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS: /datastore/sw/gentoo2/usr/lib64/blas/reference/libblas.so.3.7.0
## LAPACK: /datastore/sw/gentoo2/usr/lib64/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
##  [1] grid      stats4    parallel  stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] org.Mm.eg.db_3.7.0    topGO_2.34.0          SparseM_1.77         
##  [4] GO.db_3.7.0           AnnotationDbi_1.44.0  IRanges_2.16.0       
##  [7] S4Vectors_0.20.1      Biobase_2.42.0        graph_1.60.0         
## [10] BiocGenerics_0.28.0   dynamicTreeCut_1.63-1 UpSetR_1.3.3         
## [13] gplots_3.0.1.1        RColorBrewer_1.1-2    ggplot2_3.1.1        
## [16] edgeR_3.24.3          limma_3.38.3         
## 
## loaded via a namespace (and not attached):
##  [1] gtools_3.8.1       tidyselect_0.2.5   locfit_1.5-9.1    
##  [4] xfun_0.6           purrr_0.3.2        lattice_0.20-35   
##  [7] colorspace_1.4-1   htmltools_0.3.6    yaml_2.2.0        
## [10] blob_1.1.1         rlang_0.3.4        pillar_1.3.1      
## [13] glue_1.3.1         withr_2.1.2        DBI_1.0.0         
## [16] bit64_0.9-7        matrixStats_0.54.0 plyr_1.8.4        
## [19] stringr_1.4.0      munsell_0.5.0      gtable_0.3.0      
## [22] caTools_1.17.1.2   memoise_1.1.0      evaluate_0.13     
## [25] labeling_0.3       knitr_1.22         Rcpp_1.0.1        
## [28] KernSmooth_2.23-15 scales_1.0.0       gdata_2.18.0      
## [31] bit_1.1-14         gridExtra_2.3      digest_0.6.18     
## [34] stringi_1.4.3      dplyr_0.8.0.1      tools_3.5.1       
## [37] bitops_1.0-6       magrittr_1.5       lazyeval_0.2.2    
## [40] tibble_2.1.1       RSQLite_2.1.1      crayon_1.3.4      
## [43] pkgconfig_2.0.2    assertthat_0.2.1   rmarkdown_1.12    
## [46] R6_2.4.0           compiler_3.5.1
```

