---
title: "Hox gene expression"
date: "3 June 2019"
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---



Hox genes are critical in imparting somite identity. In mouse, there are 13 different genes (1 to 13), each with up to four paralogs (A to D). As such, there is substantial redundancy in their expression patterns and function. Hox genes form each paralogous group are clustered and ordered from 1 to 13. The first Hox genes are expressed the earliest and subsequent genes are expressed later in development. That is, Hox1 genes are the first to be expressed and Hox13 are the last.

Each somite has a specific combination of Hox gene expression, with earlier somites having predominantly expression from early Hox genes, whilst these are downregulated in later somites that instead express later Hox genes. Such combinatorial code gives a specific identity to each somite and determines the type of structures it will develop into, i.e. cervical, thoracic, lumbar, sacral or caudal derivatives. When the Hox combinatorial code is altered the *wrong* structures develop.

Little is known about the regulation of Hox gene expression and, even less, of what downstream targets are regulated by Hox factors. This dataset provides an interesting setup to explore these questions.

### Data

We have the complete transcriptome of individual mouse somites spanning early (8 somites) to late (35 somites) development. At each stage, we have sequenced the last three somites that were segmented (with somiteI being the most recent, and somiteIII the one segmented two cycles before). 


```r
# metadata
meta <- read.table(paste0(dir, "RNA-seq/data/metadata_RNAseq.tsv"), stringsAsFactors = FALSE, header = TRUE)
meta <- meta[meta$use==1,]

# raw data
data <- read.table(paste0(dir, "RNA-seq/data/geneCounts.RAW.tsv"))
y <- readRDS(paste0(dir, "RNA-seq/results/01_edgeRobject.Rds"))

# normalised data
dataNorm <- cpm(y, log = TRUE, prior.count = 1)

# remove batch effect
pcs <- read.table(paste0(dir, "RNA-seq/results/02_pcs_residuals.tab"))
dataNorm <- removeBatchEffect(dataNorm, design = model.matrix(~0+group, meta), covariates = pcs[,1:14])
dataNorm <- cbind(data[match(row.names(dataNorm), row.names(data)),1], as.data.frame(dataNorm))
colnames(dataNorm)[1] <- "gene"

table(meta$somite, stage=meta$stage)
```

```
##       stage
##        8 18 21 25 27 35
##   SI   4  5  4  4  4  6
##   SII  4  5  3  3  4  4
##   SIII 4  5  4  4  3  6
```

Given the stages we have sampled we should have a good representation of the Hox code, with representation of most, if not all, Hox genes.

### Hox gene expression

Simply plotting the relative expression of all Hox genes across the samples we see good clustering of the samples based on their stage, represented by the colour band at the top:

- green = cervical (stage8)
- light blue = thoracic (stag18)
- blue = thoracic (stage21)
- orange = lumbar (stage25)
- dark orange = lumbar (stage27)
- red = sacral (stage35)

For each gene, we have subtracted the mean expression across all samples to standardise the data.


```r
hox <- dataNorm[grep("^Hox",dataNorm$gene),]
hox <- hox[grep("s", hox$gene, invert=TRUE),] # remove antisense and non-coding genes
row.names(hox) <- hox$gene
hox$gene <- NULL

hox.cnt <- t(apply(hox, 1, function(x) x-mean(x)))
# hox.cnt <- t(apply(hox, 1, function(x) (x-min(x))/(max(x)-min(x)) ))

cols <- as.character(factor(meta$stage, labels = c("darkolivegreen4", "skyblue", "steelblue2", "darkorange", "darkorange2", "indianred1")))
hc <- hclust(dist(t(as.matrix(hox.cnt))))
order.col <- hc$order[c(1:48,65:76,49:64)]
hc <- hclust(dist(as.matrix(hox.cnt)))
order.row <- hc$order
heatmap.2(as.matrix(hox.cnt[order.row,order.col]), trace="none", col=brewer.pal(n=8, name="Blues"), ColSideColors = cols[order.col], Colv = FALSE, dendrogram = "none", labCol = rep("",ncol(hox.cnt)), Rowv = FALSE)
```

![](04.1_HoxGenes_files/figure-html/hoxGenes-1.png)<!-- -->

Cervical somite express only Hox genes 1-5 and these genes decrease in expression as development proceeds. Thoracic somites express genes 7-9. Lumbar somites maintain expression of genes 7-9 but also turn on Hox10 and the latest ones Hox11 (dark orange). Finally, sacral somites express genes 10 to 13, with 12 and 13 being specific to them.

And given that the expression of the Hox genes is so consistent with the stage, a PCA on only these 39 genes recapitulates the developmental progression of the samples.


```r
pca <- prcomp(t(hox))
df <- as.data.frame(pca$x)
ggplot(df, aes(PC1, PC2, colour=as.factor(meta$stage), shape=meta$somite)) + geom_point(cex=2) + labs(colour="stage", shape="somite") + scale_color_manual(values = c("darkolivegreen4", "skyblue", "steelblue2", "darkorange", "darkorange2", "indianred1"))
```

![](04.1_HoxGenes_files/figure-html/hox_pca-1.png)<!-- -->

### Possible targets of Hox factors

Since we have the complete transcriptome of these somites, we can identify genes that covary with Hox expression. These could be directly or indirectly regulated by Hox factors. To do this we compute the Spearman correlation between every Hox gene and all other genes expressed at moderate to high levels.


```r
stopifnot(identical(colnames(dataNorm)[-1], meta$sample))

plot(density(rowMeans(dataNorm[,-1])), main="mean expression all genes", bty="l")
abline(v=1.5, lty=2)
```

![](04.1_HoxGenes_files/figure-html/unnamed-chunk-1-1.png)<!-- -->

```r
keep <- rowMeans(dataNorm[,-1])>1.5
# summary(keep) # 10905
```

That is 10905 genes. The significance of the correlation is assessed based on a null distribution constructed from the randomised expression of each gene. 

About a third of all pairs are deemed significant.


```r
## use scran's correlatePairs function. 
## since scran cannot be installed in the server, I ran this locally and load here the results
## for reference: 
# ## only find correlations to non-HOX genes 
# tmp <- dataNorm[keep,]
# hox.idx <- which(row.names(tmp) %in% row.names(hox))
# tmp <- tmp[-hox.idx,]
# pairs <- matrix(c(rep(row.names(hox), each=nrow(tmp)), rep(row.names(tmp), nrow(hox))), ncol=2)
# 
# ## compute correlations
# tmp <- rbind(hox, tmp)
# set.seed(3957)
# corrs <- correlatePairs(as.matrix(tmp[,-1]), pairings = pairs, iters = 1e7)
corrs <- read.table(paste0(dir, "RNA-seq/results/04.1_Hox_correlatedGenes.tsv"), stringsAsFactors = FALSE, header = TRUE)
summary(corrs$FDR<0.05)
```

```
##    Mode   FALSE    TRUE 
## logical  270866  153415
```

```r
corrs.sig <- corrs[corrs$FDR<0.05,]
# nrow(corrs.sig) # 153415 significant pairs
# length(unique(corrs.sig$gene2)) # 10197 different genes 
```

This represents 10197 different non-Hox genes. Each Hox gene is involved in around four thousand significant pairs, but the variation is quite large:


```r
t <- table(corrs.sig$gene1)
names(t) <- dataNorm[match(names(t), row.names(dataNorm)),1]
plot(t[order(t, decreasing = TRUE)], las=2, ylab="significant pairs")
```

![](04.1_HoxGenes_files/figure-html/sigPairs-1.png)<!-- -->

The correlation values for the significant pairs are as low as 0.27, so there is large potential to narrow down the list of significant pairs to only strongly correlated genes.


```r
plot(density(corrs.sig$rho), main="Spearman's rho", axes=FALSE)
box(bty="l"); axis(2); axis(1, at=seq(-1,1,0.25), labels = seq(-1,1,0.25))
abline(v=c(-0.25,0.25), lty=2)
```

![](04.1_HoxGenes_files/figure-html/rho-1.png)<!-- -->

For example, if we restrict only to gene pairs with absolute correlation coefficients of 0.5 or higher the number of pairs decreases to 34425, comprising 4954 different non-Hox genes.

This affects the correlations with certain Hox genes much more than others, loosing most of the significant pairs for a number of Hox factors, and the rest being involved in a few hundred to a couple thousand gene pairs.


```r
corrs.sig.min <- corrs.sig[abs(corrs.sig$rho)>0.5,]
# dim(corrs.sig.min) # 34425
# length(unique(corrs.sig.min$gene2)) # 4954

t <- table(corrs.sig.min$gene1)
names(t) <- dataNorm[match(names(t), row.names(dataNorm)),1]
plot(t[order(t, decreasing = TRUE)], las=2, ylab="significant pairs (|rho|>0.5)")
```

![](04.1_HoxGenes_files/figure-html/minRho-1.png)<!-- -->

These ~5,000 genes correlated to Hox expression are involved in core processes such as metabolism, transcription, translation and cell cycle; less significant terms relate to development of certain organ systems.


```r
genes <- unique(corrs.sig.min$gene2)

universe <- row.names(dataNorm)
all <- as.factor(as.numeric(universe %in% genes))
names(all) <- universe

go <- new("topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl")
go.test <- runTest(go, algorithm = "classic", statistic = "Fisher" )
go.res <- GenTable(go, Fisher.classic = go.test, topNodes = length(score(go.test)))
go.res$Fisher.classic.adj <- p.adjust(go.res$Fisher.classic, "fdr")
go.res[c(1,14,18,19,20,74,79,90,128,142,153,197,199,207,208,230,268,274),]
```

```
##          GO.ID                                 Term Annotated Significant
## 1   GO:0008152                    metabolic process      8304        2734
## 14  GO:0010467                      gene expression      3980        1356
## 18  GO:0055114          oxidation-reduction process       760         312
## 19  GO:0009058                 biosynthetic process      4316        1445
## 20  GO:0022613 ribonucleoprotein complex biogenesis       393         178
## 74  GO:0006412                          translation       521         206
## 79  GO:0006325               chromatin organization       633         243
## 90  GO:0007049                           cell cycle      1436         503
## 128 GO:0008380                         RNA splicing       339         137
## 142 GO:0009790                   embryo development      1066         375
## 153 GO:0060485               mesenchyme development       256         106
## 197 GO:0007507                    heart development       558         205
## 199 GO:0048738    cardiac muscle tissue development       229          94
## 207 GO:0061515             myeloid cell development        66          34
## 208 GO:0060562        epithelial tube morphogenesis       353         136
## 230 GO:0001822                   kidney development       251         100
## 268 GO:0001837 epithelial to mesenchymal transition       126          55
## 274 GO:0007517             muscle organ development       359         135
##     Expected Fisher.classic Fisher.classic.adj
## 1    2451.51        2.2e-26       1.866920e-22
## 14   1174.98        1.1e-13       6.667571e-11
## 18    224.37        2.2e-12       1.037178e-09
## 19   1274.17        6.7e-12       2.992432e-09
## 20    116.02        1.5e-11       6.364500e-09
## 74    153.81        4.1e-07       4.701703e-05
## 79    186.87        6.2e-07       6.659899e-05
## 90    423.94        1.1e-06       1.037178e-04
## 128   100.08        9.6e-06       6.364500e-04
## 142   314.70        1.9e-05       1.135451e-03
## 153    75.58        2.9e-05       1.608458e-03
## 197   164.73        0.00011       4.738376e-03
## 199    67.61        0.00012       5.041188e-03
## 207    19.48        0.00014       5.739324e-03
## 208   104.21        0.00015       6.119712e-03
## 230    74.10        0.00027       9.961826e-03
## 268    37.20        0.00050       1.577323e-02
## 274   105.98        0.00053       1.641453e-02
```

This indicates that we are picking up many spurious correlations, that probably reflect the changes in regulation of core processes as development proceeds, instead of regulation by Hox factors. And that is the tricky part here: since Hox expression is strongly correlated with developmental correlation, it is impossible to separate the two.

However, we have matched ATAC-seq data from the same samples, so we can look for the genes within this list that have binding sites for Hox factors in their promoters, and that are open. These are much more plausible candidates for direct regulation and, maybe, more informative.

We will tackle this question in the transcription factor analysis of the ATAC-seq peaks.


```r
sessionInfo()
```

```
## R version 3.5.1 (2018-07-02)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS: /datastore/sw/gentoo2/usr/lib64/blas/reference/libblas.so.3.7.0
## LAPACK: /datastore/sw/gentoo2/usr/lib64/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] org.Mm.eg.db_3.7.0   topGO_2.34.0         SparseM_1.77        
##  [4] GO.db_3.7.0          AnnotationDbi_1.44.0 IRanges_2.16.0      
##  [7] S4Vectors_0.20.1     Biobase_2.42.0       graph_1.60.0        
## [10] BiocGenerics_0.28.0  gplots_3.0.1.1       RColorBrewer_1.1-2  
## [13] ggplot2_3.1.1        edgeR_3.24.3         limma_3.38.3        
## 
## loaded via a namespace (and not attached):
##  [1] gtools_3.8.1       tidyselect_0.2.5   locfit_1.5-9.1    
##  [4] xfun_0.6           purrr_0.3.2        lattice_0.20-35   
##  [7] colorspace_1.4-1   htmltools_0.3.6    yaml_2.2.0        
## [10] blob_1.1.1         rlang_0.3.4        pillar_1.3.1      
## [13] glue_1.3.1         withr_2.1.2        DBI_1.0.0         
## [16] bit64_0.9-7        matrixStats_0.54.0 plyr_1.8.4        
## [19] stringr_1.4.0      munsell_0.5.0      gtable_0.3.0      
## [22] caTools_1.17.1.2   evaluate_0.13      memoise_1.1.0     
## [25] labeling_0.3       knitr_1.22         Rcpp_1.0.1        
## [28] KernSmooth_2.23-15 scales_1.0.0       gdata_2.18.0      
## [31] bit_1.1-14         digest_0.6.18      stringi_1.4.3     
## [34] dplyr_0.8.0.1      grid_3.5.1         tools_3.5.1       
## [37] bitops_1.0-6       magrittr_1.5       lazyeval_0.2.2    
## [40] tibble_2.1.1       RSQLite_2.1.1      crayon_1.3.4      
## [43] pkgconfig_2.0.2    assertthat_0.2.1   rmarkdown_1.12    
## [46] R6_2.4.0           compiler_3.5.1
```
