### Control RNA-seq data

Before producing the paired RNA- and ATAC-seq dataset it was important to determine that the left and right somites are indeed equivalent and do not show any significant differnces in expression.

To this end we sequenced four pairs of somites from two different embryos from the same litter (and thus roughly from the same somite stage).

Additionally, for one pair of somites, we repeated the library prep procedure using only one quarter of the recommended reagents quantity. If the two libraries are quivalent we could then save on reagents by four-fold!

These libraries were sequenced on one lane of an Illumina HiSeq4000, to produce single-end 50bp fragments.

#### Mapping the sequencing data

Since the libraries were sequenced on one lane only there is no need to merge FASTq files. Instead, we create symbolic links to the files in the Odom Lab repository.

```{bash}
ln -s /ark03/repository/do16433/do16433_unk_somite_SI-1_unk_mm10BL6e1_CRI01.fq.gz /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/FASTq/do16433.fq.gz
ln -s /ark03/repository/do16434/do16434_unk_somite_SI-2_unk_mm10BL6e1_CRI01.fq.gz /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/FASTq/do16434.fq.gz
ln -s /ark03/repository/do16435/do16435_unk_somite_SII-1_unk_mm10BL6e1_CRI01.fq.gz /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/FASTq/do16435.fq.gz
ln -s /ark03/repository/do16436/do16436_unk_somite_SII-2_unk_mm10BL6e1_CRI01.fq.gz /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/FASTq/do16436.fq.gz
ln -s /ark03/repository/do16437/do16437_unk_somite_SI-1_unk_mm10BL6e2_CRI01.fq.gz /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/FASTq/do16437.fq.gz
ln -s /ark03/repository/do16438/do16438_unk_somite_SI-2_unk_mm10BL6e2_CRI01.fq.gz /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/FASTq/do16438.fq.gz
ln -s /ark03/repository/do16439/do16439_unk_somite_SII-1_unk_mm10BL6e2_CRI01.fq.gz /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/FASTq/do16439.fq.gz
ln -s /ark03/repository/do16440/do16440_unk_somite_SII-2_unk_mm10BL6e2_CRI01.fq.gz /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/FASTq/do16440.fq.gz
ln -s /ark03/repository/do16441/do16441_unk_somite_SII-1_unk_mm10BL6e1_CRI01.fq.gz /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/FASTq/do16441.fq.gz
ln -s /ark03/repository/do16442/do16442_unk_somite_SII-2_unk_mm10BL6e1_CRI01.fq.gz /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/FASTq/do16442.fq.gz
```

We can then directly proceed to map the sequencing reads to the mouse genome, following the same method as with the main dataset (see `01.0_mapping_RNAseq_reads.md` for details).

We map:

```{bash}
sh 02.1_align.sh
# 35980
```

Compute mapping statistics:

```{bash}
perl 01.3_mappingStats.pl -dir /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/STAR/CONTROL/ > /user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/RNA-seq/data/mappingStatistics_CONTROLlibraries.tsv
```

And create a counts matrix:

```{bash}
R CMD BATCH 02.2_countMatrix.R
```
