---
title: "Figure S1"
date: "02 July 2019"
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(Rsamtools)
library(csaw)
library(ggplot2)
library(zoo)
library(UpSetR)

dir <- "/user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/"

th <- theme_bw() + theme(axis.text.x = element_text(size=10), axis.title.x = element_text(size=10), axis.text.y = element_text(size=10), axis.title.y = element_text(size=10), panel.grid.major = element_blank(), panel.grid.minor = element_blank(), axis.line = element_line(colour = "black"), panel.border = element_blank(), plot.title = element_text(face="bold", hjust = 0.5, size=10))

#### FUNCTIONS
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)
  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)
  numPlots = length(plots)
  
  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                     ncol = cols, nrow = ceiling(numPlots/cols), byrow = TRUE)
  }
  
  if (numPlots==1) {
    print(plots[[1]])
    
  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
    
    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
      
      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}
```

### Figure S1

This figure exemplifies the criteria used to QC the ATAC-seq data.

```{r meta}
meta <- read.table(paste0(dir, "ATAC-seq/data/metadata_ATACseq.tsv"), stringsAsFactors = FALSE, header = TRUE)
```

#### Insert size distribution

```{r insSize, fig.width=12, fig.height=2.5}
## in 01_insertSizeDistribution.Rmd we obtain the fragmen sizes for each clean BAM
diagnostics <- readRDS(paste0(dir, "ATAC-seq/results/01_mappingStats_ATACseq.Rds"))
stopifnot(identical(substr(names(diagnostics), 1, nchar(names(diagnostics))-14), meta$sample))

## get a representative sample from each insert size score group
samples <- c("e23_SII-2", "e16_SI-2", "e26_SII-2", "e32_SII-2", "e14_SII-2")

# pdf(paste0(dir, "Figures/PDFs/FigureS1_insSizeDist.pdf"), width = 15, height = 3)
par(mfrow=c(1,5), mar=c(2,2,2,2))
for(i in 0:4){
  plot(density(diagnostics[[grep(samples[i+1], names(diagnostics))]]$sizes), axes=FALSE, main=paste("score =",i), xlab="", ylab="", lwd=3, xlim=c(0,1000))
  box(bty="l"); axis(1)
}
# dev.off()
```

```{r corrs, fig.width=9, fig.height=3}
plots <- list()
plots[[1]] <- ggplot(meta, aes(as.factor(insSizeDist), size)) + geom_boxplot() + ggtitle("DNA fragment size") + xlab("insert size dist score") + ylab("bp") + th
plots[[2]] <- ggplot(meta, aes(as.factor(insSizeDist), log10(conc))) + geom_boxplot() + ggtitle("DNA concentration") + xlab("insert size dist score") + ylab(expression('log'[10]*' [nM]')) + th
plots[[3]] <- ggplot(meta, aes(as.factor(insSizeDist), goodQuality/1e6)) + geom_boxplot() + ggtitle("Total unique GQ alignments") + xlab("insert size dist score") + ylab("million") + th

# postscript(paste0(dir, "Figures/PDFs/FigureS1_insSizeDist_corrs.eps"), width = 12, height = 4)
multiplot(plotlist = plots, cols=3)
# dev.off()
```


#### Peak calling

```{r peakCalls, fig.width=8, fig.height=4, message=FALSE}
data_summary <- function(x) {
   m <- mean(x)
   ymin <- m-sd(x)
   ymax <- m+sd(x)
   return(c(y=m,ymin=ymin,ymax=ymax))
}

plots <- list()
plots[[1]] <- ggplot(meta, aes(as.factor(insSizeDist), nPeaks/1e3)) + geom_violin() + geom_jitter(width=0.15) + xlab("insert size dist score") + ylab("number of peaks x 1000") + stat_summary(fun.data=data_summary, col="red") + th
plots[[2]] <- ggplot(meta, aes(as.factor(insSizeDist), readsInPeaks/goodQuality*100)) + geom_violin() + geom_jitter(width=0.15) +  xlab("insert size dist score") + ylab("FRiP") + stat_summary(fun.data=data_summary, col="red") + th

# plots <- list()
# plots[[1]] <- ggplot(meta, aes(as.factor(insSizeDist), nPeaks/1e3)) + geom_violin() + geom_jitter(width=0.15) + xlab("insert size dist score") + ylab("number of peaks x 1000") + stat_summary(fun.data=data_summary, col="red") + th + geom_hline(yintercept = c(2.652, 5.273, 23.985, 35.831, 57.994))
# plots[[2]] <- ggplot(meta, aes(as.factor(insSizeDist), readsInPeaks/goodQuality*100)) + geom_violin() + geom_jitter(width=0.15) +  xlab("insert size dist score") + ylab("FRiP") + stat_summary(fun.data=data_summary, col="red") + th + geom_hline(yintercept = c(0.3089941, 0.9290860, 4.5579091, 7.9610104, 19.8634197))

# postscript(paste0(dir, "Figures/PDFs/FigureS1_numberPeaks.eps"), width = 8, height = 4)
multiplot(plotlist = plots, cols=2)
# dev.off()
# geom_dotplot(binaxis='y', stackdir='center', dotsize=0.7)
```

#### TSS enrichment

```{r tss, fig.width=12, fig.height=2.5}
tss <- readRDS(paste0(dir, "ATAC-seq/results/02_TSSinsertionCounts.Rds"))
tss.norm <- t(do.call("cbind", lapply(tss, function(x) colMeans(x)/mean(colMeans(x[,c(1:100,1901:2001)])))))

# pdf(paste0(dir, "Figures/PDFs/FigureS1_TSSenrich.pdf"), width = 15, height = 3)
par(mfrow=c(1,5), mar=c(2,2,2,2))
for(sample in samples){
  plot(rollmean(tss.norm[sample,], k=25), type="l", lwd=3, main=sample, xlab="", ylab="", ylim=c(1,7), axes=FALSE)
  box(bty="l"); axis(1, at=c(0,1000,2000), labels = c("-1kb","TSS","1kb")); axis(2,las=2)
  abline(h=4, lty=2, lwd=2)
}
# dev.off()
```

#### Quality control

```{r qc}
qc <- data.frame(nuclosome = ifelse(meta$insSizeDist>=2, 1, 0), nPeaks = ifelse(meta$nPeaks > 15000, 1, 0), frip = ifelse(meta$readsInPeaks/meta$goodQuality*100 >= 3, 1, 0), tss = ifelse(meta$TSSscore > 4, 1, 0))
row.names(qc) <- meta$sample

# pdf(paste0(dir, "Figures/PDFs/FigureS1_upsetPlot.pdf"), width = 5, height = 5)
upset(qc, mainbar.y.label = "number of samples", sets.x.label = "number of samples\nthat pass", text.scale=1.25)
# dev.off()
```







```{r info}
sessionInfo()
```

