---
title: "Differential accessibility analysis"
date: "05 April 2019"
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(csaw)
library(edgeR)
library(ChIPpeakAnno)

dir <- "/user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/"
```

We have ATAC-seq data for 50 good-quality samples, comprising the last three generated somites of mouse embryos at six different developmental stages.

```{r data}
## metadata
meta <- read.table(paste0(dir, "ATAC-seq/data/metadata_ATACseq.tsv"), stringsAsFactors = FALSE, header = TRUE)
meta <- meta[meta$QCpass==1,]

## data
filtered.data <- readRDS(paste0(dir, "ATAC-seq/results/03_windowCounts_filteredWindows.Rds"))
sf.trended <- readRDS(paste0(dir, "ATAC-seq/results/03_trendedNorm_sizeFactors.Rds"))

norm.counts <- log2(assay(filtered.data)+0.5) - sf.trended/log(2)
```


We have defined a set of peaks representing regions of open chromatin, and computed the sequencing data counts for 150bp windows covering such loci. We have used a trended-normalisation approach to remove systematic biases. On this normalised dataset, the major source of variation was still strongly correlated to the fraction of reads in peaks (FRiP) for each sample. Thus, we have tried two approaches to capture such variation:

* Surrogate variable analysis: identifies one surrogate variable that is strongly correlated to FRiP and thus should be able to remove this effect.

* PCA on the residuals of the model containing the somite-stage information: the first few PCs are correlated to FRiP and consideration of the first 18 components cleans up the data nicely.

As a baseline, we will perform the differential analysis without any correction, and assuming the normalisation has successfully removed technical biases. We can then compare these results to those obtained when incorporating the SVA or PCA covariates, to see if they are indeed removing technical noise or instead are capturing biological signal. 

### Differential accessibility analysis

We can test for differences between the somite trios and between stages.

We use `edgeR` to perform the differential testing. We first estimate the dispersion and fit the design containing the interaction of somite and stage.

```{r dispersionFit, fig.width=10}
## edgeR object
y <- asDGEList(filtered.data)
y$offset <- sf.trended

## design
meta$group <- factor(paste(meta$stage, meta$somite, sep="."))
design <- model.matrix(~0+group, meta)
colnames(design) <- paste0("stage",levels(meta$group))

## dispersion
y <- estimateDisp(y, design)
par(mfrow=c(1,2))
plotBCV(y)

## fit model
fit <- glmQLFit(y, design, robust=TRUE)
plotQLDisp(fit)
# summary(fit$df.prior)

saveRDS(y, file=paste0(dir, "ATAC-seq/results/04.1_edgeRobject.Rds"))
```

##### Differences across somite trios

We can now compare the chromatin accessibility as somites differentiate (somite I vs II vs III). First, we perform the comparison for each stage. We are underpowered in several stages, where we don't have replicates for a particular somite-stage combination.

```{r reps}
table(meta$stage, meta$somite)
```

We definitely exclude stage27 from the analysis. For all other stages, we define all three pairwise comparisons, and test them all together with an anova-like test.

```{r somiteTrios_perStage}
my.contrasts <- makeContrasts(stage8.IIvsI=stage8.SII-stage8.SI, stage8.IIIvsI=stage8.SIII-stage8.SI, 
                              stage8.IIIvsII=stage8.SIII-stage8.SII, 
                              stage18.IIvsI=stage18.SII-stage18.SI, stage18.IIIvsI=stage18.SIII-stage18.SI, 
                              stage18.IIIvsII=stage18.SIII-stage18.SII, 
                              stage21.IIvsI=stage21.SII-stage21.SI, stage21.IIIvsI=stage21.SIII-stage21.SI, 
                              stage21.IIIvsII=stage21.SIII-stage21.SII, 
                              stage25.IIvsI=stage25.SII-stage25.SI, stage25.IIIvsI=stage25.SIII-stage25.SI, 
                              stage25.IIIvsII=stage25.SIII-stage25.SII, 
                              stage35.IIvsI=stage35.SII-stage35.SI, stage35.IIIvsI=stage35.SIII-stage35.SI, 
                              stage35.IIIvsII=stage35.SIII-stage35.SII, levels=design)

## test
stage8 <- glmQLFTest(fit, contrast = my.contrasts[,1:3])
stage18 <- glmQLFTest(fit, contrast = my.contrasts[,4:6])
stage21 <- glmQLFTest(fit, contrast = my.contrasts[,7:9])
stage25 <- glmQLFTest(fit, contrast = my.contrasts[,10:12])
stage35 <- glmQLFTest(fit, contrast = my.contrasts[,13:15])
```

The tests are performed on every window that passed the minimum abundance filter and is within the peak set. However, we are interested in regions of open chromatin as a whole -i.e. *peaks*-, more than the individual windows *per se*. Thus, we merge all adjacent windows that are no more than 150bp apart into regions. We also set an upper limit of 1kb to avoid chaining reactions of broadly open areas where a single region could be very long. In these cases, the locus simply gets split into adjacent regions of 1.5kb. For each defined region, we compute a combined p-value against the null hypothesis that none of the windows are differentially accessible.

Not surprisingly, there are no significant findings for most of the stages. Only the comparison for stage 25 returns a significant number of differential regions, probably because of the larger number of replicates and higher sample quality (see `02_qualityControl.md`).

```{r somiteTrios_perStage_merge}
merged <- mergeWindows(rowRanges(filtered.data), tol=150, max.width = 1500)

somiteTrios.perStage <- list()
somiteTrios.perStage[['stage8']] <- merged$region
mcols(somiteTrios.perStage[['stage8']]) <- combineTests(merged$id, stage8$table)
somiteTrios.perStage[['stage18']] <- merged$region
mcols(somiteTrios.perStage[['stage18']]) <- combineTests(merged$id, stage18$table)
somiteTrios.perStage[['stage21']] <- merged$region
mcols(somiteTrios.perStage[['stage21']]) <- combineTests(merged$id, stage21$table)
somiteTrios.perStage[['stage25']] <- merged$region
mcols(somiteTrios.perStage[['stage25']]) <- combineTests(merged$id, stage25$table)
somiteTrios.perStage[['stage35']] <- merged$region
mcols(somiteTrios.perStage[['stage35']]) <- combineTests(merged$id, stage35$table)
saveRDS(somiteTrios.perStage, paste0(dir, "ATAC-seq/results/04.1_diffAccessibility_somiteTrios_perStage.Rds"))

df <- t(data.frame(stage8=sum(mcols(somiteTrios.perStage[['stage8']])$FDR <= 0.05),
           stage18=sum(mcols(somiteTrios.perStage[['stage18']])$FDR <= 0.05), 
           stage21=sum(mcols(somiteTrios.perStage[['stage21']])$FDR <= 0.05),
           stage25=sum(mcols(somiteTrios.perStage[['stage25']])$FDR <= 0.05),
           stage35=sum(mcols(somiteTrios.perStage[['stage35']])$FDR <= 0.05)))
colnames(df) <- "nDA"
df
```

To increase power we combine samples from all stages and test the average of each somite. We will loose any changes that are stage-specific but should recover more of the consistent changes:

```{r somiteTrios_average}
my.contrasts <- makeContrasts(
  somiteIIvsI = (stage8.SII + stage18.SII + stage21.SII + stage25.SII + stage27.SII + stage35.SII)/6 - (stage8.SI + stage18.SI + stage21.SI + stage25.SI + stage27.SI + stage35.SI)/6,
  somiteIIIvsI = (stage8.SIII + stage18.SIII + stage21.SIII + stage25.SIII + stage27.SIII + stage35.SIII)/6 - (stage8.SI + stage18.SI + stage21.SI + stage25.SI + stage27.SI + stage35.SI)/6,
  somiteIIIvsII = (stage8.SIII + stage18.SIII + stage21.SIII + stage25.SIII + stage27.SIII + stage35.SIII)/6 - (stage8.SII + stage18.SII + stage21.SII + stage25.SII + stage27.SII + stage35.SII)/6,
  levels=design)

somiteIIvsI.all <- glmQLFTest(fit, contrast = my.contrasts[,"somiteIIvsI"])
somiteIIIvsI.all <- glmQLFTest(fit, contrast = my.contrasts[,"somiteIIIvsI"])
somiteIIIvsII.all <- glmQLFTest(fit, contrast = my.contrasts[,"somiteIIIvsII"])

## data.frames
somiteTrios.average <- list()

somiteTrios.average[['somiteIIvsI']] <- merged$region
mcols(somiteTrios.average[['somiteIIvsI']]) <- combineTests(merged$id, somiteIIvsI.all$table)
somiteTrios.average[['somiteIIvsI']] <- somiteTrios.average[['somiteIIvsI']][order(somiteTrios.average[['somiteIIvsI']]$FDR)]
somiteTrios.average[['somiteIIIvsI']] <- merged$region
mcols(somiteTrios.average[['somiteIIIvsI']]) <- combineTests(merged$id, somiteIIIvsI.all$table)
somiteTrios.average[['somiteIIIvsI']] <- somiteTrios.average[['somiteIIIvsI']][order(somiteTrios.average[['somiteIIIvsI']]$FDR)]
somiteTrios.average[['somiteIIIvsII']] <- merged$region
mcols(somiteTrios.average[['somiteIIIvsII']]) <- combineTests(merged$id, somiteIIIvsII.all$table)
somiteTrios.average[['somiteIIIvsII']] <- somiteTrios.average[['somiteIIIvsII']][order(somiteTrios.average[['somiteIIIvsII']]$FDR)]

saveRDS(somiteTrios.average, paste0(dir, "ATAC-seq/results/04.1_diffAccessibility_somiteTrios_average.Rds"))

data.frame(somiteIvsII=sum(mcols(somiteTrios.average[['somiteIIvsI']])$FDR <= 0.05),
           somiteIvsIII=sum(mcols(somiteTrios.average[['somiteIIIvsI']])$FDR <= 0.05), 
           somiteIIvsIII=sum(mcols(somiteTrios.average[['somiteIIIvsII']])$FDR <= 0.05))
```

Again, the numbers are low, but a few hundred differences are picked up between somites I and III, which are the most different between them.

##### Differences across developmental stages

Now, we can also test for changes across development. First, we test all pairwise comparisons between stages on a per-somite basis. 

```{r stage_perSomite}
my.contrasts <- makeContrasts(stage8vs18 = stage18.SI - stage8.SI, stage8vs21 = stage21.SI - stage8.SI,
                              stage8vs25 = stage25.SI - stage8.SI, stage8vs27 = stage27.SI - stage8.SI, 
                              stage8vs35 = stage35.SI - stage8.SI, 
                              stage18vs21 = stage21.SI - stage18.SI, stage18vs25 = stage25.SI - stage18.SI, 
                              stage18vs27 = stage27.SI - stage18.SI, stage18vs35 = stage35.SI - stage18.SI, 
                              stage21vs25 = stage25.SI - stage21.SI, stage21vs27 = stage27.SI - stage21.SI, 
                              stage21vs35 = stage35.SI - stage21.SI, 
                              stage25vs27 = stage27.SI - stage25.SI, 
                              stage27vs35 = stage27.SI - stage35.SI, levels=design)

somiteI <- glmQLFTest(fit, contrast = my.contrasts)

my.contrasts <- makeContrasts(stage8vs18 = stage18.SII - stage8.SII, stage8vs21 = stage21.SII - stage8.SII,
                              stage8vs25 = stage25.SII - stage8.SII, stage8vs27 = stage27.SII - stage8.SII, 
                              stage8vs35 = stage35.SII - stage8.SII, 
                              stage18vs21 = stage21.SII - stage18.SII, stage18vs25 = stage25.SII - stage18.SII, 
                              stage18vs27 = stage27.SII - stage18.SII, stage18vs35 = stage35.SII - stage18.SII, 
                              stage21vs25 = stage25.SII - stage21.SII, stage21vs27 = stage27.SII - stage21.SII, 
                              stage21vs35 = stage35.SII - stage21.SII, 
                              stage25vs27 = stage27.SII - stage25.SII, 
                              stage27vs35 = stage27.SII - stage35.SII, levels=design)
somiteII <- glmQLFTest(fit, contrast = my.contrasts)

my.contrasts <- makeContrasts(stage8vs18 = stage18.SIII - stage8.SIII, stage8vs21 = stage21.SIII - stage8.SIII,
                              stage8vs25 = stage25.SIII - stage8.SIII, stage8vs27 = stage27.SIII - stage8.SIII, 
                              stage8vs35 = stage35.SIII - stage8.SIII, 
                              stage18vs21 = stage21.SIII - stage18.SIII, stage18vs25 = stage25.SIII - stage18.SIII, 
                              stage18vs27 = stage27.SIII - stage18.SIII, stage18vs35 = stage35.SIII - stage18.SIII, 
                              stage21vs25 = stage25.SIII - stage21.SIII, stage21vs27 = stage27.SIII - stage21.SIII, 
                              stage21vs35 = stage35.SIII - stage21.SIII, 
                              stage25vs27 = stage27.SIII - stage25.SIII, 
                              stage27vs35 = stage27.SIII - stage35.SIII, levels=design)
somiteIII <- glmQLFTest(fit, contrast = my.contrasts)
```

This gives us a few thousand significant regions, with many more recovered with somiteI data. This might reflect the higher number of replicates per stage for this somite, compared to the other two. 

```{r stage_perSomite_merge}
stage.perSomite <- list()
stage.perSomite[['somiteI']] <- merged$region
mcols(stage.perSomite[['somiteI']]) <- combineTests(merged$id, somiteI$table)
stage.perSomite[['somiteII']] <- merged$region
mcols(stage.perSomite[['somiteII']]) <- combineTests(merged$id, somiteII$table)
stage.perSomite[['somiteIII']] <- merged$region
mcols(stage.perSomite[['somiteIII']]) <- combineTests(merged$id, somiteIII$table)
saveRDS(stage.perSomite, paste0(dir, "ATAC-seq/results/04.1_diffAccessibility_stages_perSomite.Rds"))

df <- t(data.frame(somiteI=sum(mcols(stage.perSomite[['somiteI']])$FDR <= 0.05),
           somiteII=sum(mcols(stage.perSomite[['somiteII']])$FDR <= 0.05), 
           somiteIII=sum(mcols(stage.perSomite[['somiteIII']])$FDR <= 0.05)))
colnames(df) <- "nDA"
df
```

Interestingly, there are plenty of regions that are significant in a somite-specific manner.

```{r overlap, message=FALSE, warning=FALSE}
p <- makeVennDiagram(list(somiteI=stage.perSomite[["somiteI"]][stage.perSomite[["somiteI"]]$FDR < 0.05,],
                          somiteII=stage.perSomite[["somiteII"]][stage.perSomite[["somiteII"]]$FDR < 0.05,],
                          somiteIII=stage.perSomite[["somiteIII"]][stage.perSomite[["somiteIII"]]$FDR < 0.05,]))
```

Now, we instead take the average of somites I, II and III. This should increase power for changes that are conserved across the AP axis, but driven by developmental progression.

```{r stage_average}
my.contrasts <- makeContrasts(stage8vs18 = (stage18.SIII+stage18.SII+stage18.SI)/3 - (stage8.SIII+stage8.SII+stage8.SI)/3,
                              stage8vs21 = (stage21.SIII+stage21.SII+stage21.SI)/3 - (stage8.SIII+stage8.SII+stage8.SI)/3,
                              stage8vs25 = (stage25.SIII+stage25.SII+stage25.SI)/3 - (stage8.SIII+stage8.SII+stage8.SI)/3,
                              stage8vs27 = (stage27.SIII+stage27.SII+stage27.SI)/3 - (stage8.SIII+stage8.SII+stage8.SI)/3,
                              stage8vs35 = (stage35.SIII+stage35.SII+stage35.SI)/3 - (stage8.SIII+stage8.SII+stage8.SI)/3,
                              stage18vs21 = (stage21.SIII+stage21.SII+stage21.SI)/3 - (stage18.SIII+stage18.SII+stage18.SI)/3,
                              stage18vs25 = (stage25.SIII+stage25.SII+stage25.SI)/3 - (stage18.SIII+stage18.SII+stage18.SI)/3,
                              stage18vs27 = (stage27.SIII+stage27.SII+stage27.SI)/3 - (stage18.SIII+stage18.SII+stage18.SI)/3,
                              stage18vs35 = (stage35.SIII+stage35.SII+stage35.SI)/3 - (stage18.SIII+stage18.SII+stage18.SI)/3,
                              stage21vs25 = (stage25.SIII+stage25.SII+stage25.SI)/3 - (stage21.SIII+stage21.SII+stage21.SI)/3,
                              stage21vs27 = (stage27.SIII+stage27.SII+stage27.SI)/3 - (stage21.SIII+stage21.SII+stage21.SI)/3,
                              stage21vs35 = (stage35.SIII+stage35.SII+stage35.SI)/3 - (stage21.SIII+stage21.SII+stage21.SI)/3,
                              stage25vs27 = (stage27.SIII+stage27.SII+stage27.SI)/3 - (stage25.SIII+stage25.SII+stage25.SI)/3,
                              stage25vs35 = (stage35.SIII+stage35.SII+stage35.SI)/3 - (stage25.SIII+stage25.SII+stage25.SI)/3,
                              stage27vs35 = (stage35.SIII+stage35.SII+stage35.SI)/3 - (stage27.SIII+stage27.SII+stage27.SI)/3, levels=design)

test <- glmQLFTest(fit, contrast = my.contrasts)
```

And we test against the null that none of the stages are different. Therefore, anything with an FDR of 5% or lower indicates a region of open chromatin that is differentially accessible between at least a pair of stages. 

Nearly 20 thousand regions are differentially accessible across development.

```{r stage_average_merge}
stage.average <- merged$region
mcols(stage.average) <- combineTests(merged$id, test$table)

summary(stage.average$FDR<0.05)

saveRDS(stage.average, paste0(dir, "ATAC-seq/results/04.1_diffAccessibility_stages_average.Rds"))
```

```{r info}
sessionInfo()
```

