---
title: "Differential accessibility analysis - PCA"
date: "05 April 2019"
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---



We have ATAC-seq data for 50 good-quality samples, comprising the last three generated somites of mouse embryos at six different developmental stages.


```r
## metadata
meta <- read.table(paste0(dir, "ATAC-seq/data/metadata_ATACseq.tsv"), stringsAsFactors = FALSE, header = TRUE)
meta <- meta[meta$QCpass==1,]

## data
filtered.data <- readRDS(paste0(dir, "ATAC-seq/results/03_windowCounts_filteredWindows.Rds"))
sf.trended <- readRDS(paste0(dir, "ATAC-seq/results/03_trendedNorm_sizeFactors.Rds"))

norm.counts <- log2(assay(filtered.data)+0.5) - sf.trended/log(2)
```

We have defined a set of peaks representing regions of open chromatin, and computed the sequencing data counts for 150bp windows covering such loci. We have used a trended-normalisation approach to remove systematic biases. On this normalised dataset, the major source of variation was still strongly correlated to the fraction of reads in peaks (FRiP) for each sample. Thus, we have tried two approaches to capture such variation:

* Surrogate variable analysis: identifies one surrogate variable that is strongly correlated to FRiP and thus should be able to remove this effect.

* PCA on the residuals of the model containing the somite-stage information: the first few PCs are correlated to FRiP and consideration of the first 18 components cleans up the data nicely.

### Differential accessibility analysis

Now we can perform differential analysis to obtain differences between the somite trios, and between stages, using the normalised data while regressing out either the SV or PCs that capture the remaining technical variation in the data.

#### PCA approach

Here, let's use as covariates the 18 PCs from the fit residuals in the design matrix. We use `edgeR` to perform the differential testing. We first estimate the dispersion and fit the design containing the interaction of somite and stage, plus the 18 PCs.


```r
## PCA on the fit residuals
pcs <- read.table(paste0(dir, "ATAC-seq/results/03_pcs_residuals.tab"))

## edgeR object
y <- asDGEList(filtered.data)
y$offset <- sf.trended

## design
meta$group <- factor(paste(meta$stage, meta$somite, sep="."))
design <- model.matrix(~0+group, meta)
colnames(design) <- paste0("stage",levels(meta$group))

design <- cbind(design, pcs[,1:18])
colnames(design)[-c(1:18)] <- paste0("PC",1:18)

## dispersion
y <- estimateDisp(y, design)
par(mfrow=c(1,2))
plotBCV(y)

## fit model
fit <- glmQLFit(y, design, robust=TRUE)
plotQLDisp(fit)
```

![](04.3_differentialAccessibility_PCA_files/figure-html/dispersionFit-1.png)<!-- -->

```r
# summary(fit$df.prior)

saveRDS(y, file=paste0(dir, "ATAC-seq/results/04.3_edgeRobject_pca.Rds"))
```

##### Differences across somite trios

We can now compare the chromatin accessibility as somites differentiate (somite I vs II vs III). First, we perform the comparison for each stage. We are underpowered in several stages, where we don't have replicates for a particular somite-stage combination.


```r
table(meta$stage, meta$somite)
```

```
##     
##      SI SII SIII
##   8   2   2    2
##   18  3   4    3
##   21  2   4    1
##   25  4   4    3
##   27  3   1    1
##   35  5   4    2
```

We definitely exclude stage27 from the analysis. For all other stages, we define all three pairwise comparisons, and test them all together with an anova-like test.


```r
my.contrasts <- makeContrasts(stage8.IIvsI=stage8.SII-stage8.SI, stage8.IIIvsI=stage8.SIII-stage8.SI, 
                              stage8.IIIvsII=stage8.SIII-stage8.SII, 
                              stage18.IIvsI=stage18.SII-stage18.SI, stage18.IIIvsI=stage18.SIII-stage18.SI, 
                              stage18.IIIvsII=stage18.SIII-stage18.SII, 
                              stage21.IIvsI=stage21.SII-stage21.SI, stage21.IIIvsI=stage21.SIII-stage21.SI, 
                              stage21.IIIvsII=stage21.SIII-stage21.SII, 
                              stage25.IIvsI=stage25.SII-stage25.SI, stage25.IIIvsI=stage25.SIII-stage25.SI, 
                              stage25.IIIvsII=stage25.SIII-stage25.SII, 
                              stage35.IIvsI=stage35.SII-stage35.SI, stage35.IIIvsI=stage35.SIII-stage35.SI, 
                              stage35.IIIvsII=stage35.SIII-stage35.SII, levels=design)

## test
trios.perStage <- list()
trios.perStage[["stage8"]] <- glmQLFTest(fit, contrast = my.contrasts[,1:3])
trios.perStage[["stage18"]] <- glmQLFTest(fit, contrast = my.contrasts[,4:6])
trios.perStage[["stage21"]] <- glmQLFTest(fit, contrast = my.contrasts[,7:9])
trios.perStage[["stage25"]] <- glmQLFTest(fit, contrast = my.contrasts[,10:12])
trios.perStage[["stage35"]] <- glmQLFTest(fit, contrast = my.contrasts[,13:15])
```

The tests are performed on every window that passed the minimum abundance filter and is within the peak set. However, we are interested in regions of open chromatin as a whole -i.e. *peaks*-, more than the individual windows *per se*. Thus, we merge all adjacent windows that are no more than 150bp apart into regions. We also set an upper limit of 1kb to avoid chaining reactions of broadly open areas where a single region could be very long. In these cases, the locus simply gets split into adjacent regions of 1.5kb. For each defined region, we compute a combined p-value against the null hypothesis that none of the windows are differentially accessible.

There is great variability in the number of regions that reach significance for the different stages. This likely reflects the number of replicates and the quality of the samples. Not surprisingly, stage21 is underpowerd; stage35 should do better but the samples from SIII are not the best quality.

Samples from stage25 were of particularly good quality and, together with lots of replicates, return the most significant regions. It is likely that many of these are also differential in the other stages, but do not reach significance. 


```r
merged <- mergeWindows(rowRanges(filtered.data), tol=150, max.width = 1500)

somiteTrios.perStage.pca <- list()
for(contrast in paste0("stage",c(8,18,21,25,35))){
  somiteTrios.perStage.pca[[contrast]] <- merged$region
  mcols(somiteTrios.perStage.pca[[contrast]]) <- combineTests(merged$id, trios.perStage[[contrast]]$table)
  best <- getBestTest(merged$id, trios.perStage[[contrast]]$table)
  mcols(somiteTrios.perStage.pca[[contrast]]) <- cbind(mcols(somiteTrios.perStage.pca[[contrast]]), best[,-c((ncol(best)-2):ncol(best))])
}
saveRDS(somiteTrios.perStage.pca, paste0(dir, "ATAC-seq/results/04.3_diffAccessibility_somiteTrios_perStage_pca.Rds"))

df <- t(data.frame(stage8=sum(mcols(somiteTrios.perStage.pca[['stage8']])$FDR <= 0.05),
           stage18=sum(mcols(somiteTrios.perStage.pca[['stage18']])$FDR <= 0.05), 
           stage21=sum(mcols(somiteTrios.perStage.pca[['stage21']])$FDR <= 0.05),
           stage25=sum(mcols(somiteTrios.perStage.pca[['stage25']])$FDR <= 0.05),
           stage35=sum(mcols(somiteTrios.perStage.pca[['stage35']])$FDR <= 0.05)))
colnames(df) <- "nDA"
df
```

```
##          nDA
## stage8   573
## stage18  193
## stage21    6
## stage25 1094
## stage35   46
```

Interestingly, the majority (80-92%) of the regions significant in stages 8, 18 or 35 are not significant in stage25, raising the possibility that there are stage specific changes.


```r
s8 <- somiteTrios.perStage.pca[["stage8"]][somiteTrios.perStage.pca[["stage8"]]$FDR<0.05]
s18 <- somiteTrios.perStage.pca[["stage18"]][somiteTrios.perStage.pca[["stage18"]]$FDR<0.05]
s25 <- somiteTrios.perStage.pca[["stage25"]][somiteTrios.perStage.pca[["stage25"]]$FDR<0.05]
s35 <- somiteTrios.perStage.pca[["stage35"]][somiteTrios.perStage.pca[["stage35"]]$FDR<0.05]

# length(unique(queryHits(findOverlaps(s8, s25))))/length(s8)
# length(unique(queryHits(findOverlaps(s18, s25))))/length(s18)
# length(unique(queryHits(findOverlaps(s35, s25))))/length(s35)
```

We can also test the average across stages of each somite. We will loose any changes that are stage-specific but should recover more of the consistent changes:


```r
my.contrasts <- makeContrasts(
  somiteIIvsI = (stage8.SII + stage18.SII + stage21.SII + stage25.SII + stage27.SII + stage35.SII)/6 - (stage8.SI + stage18.SI + stage21.SI + stage25.SI + stage27.SI + stage35.SI)/6,
  somiteIIIvsI = (stage8.SIII + stage18.SIII + stage21.SIII + stage25.SIII + stage27.SIII + stage35.SIII)/6 - (stage8.SI + stage18.SI + stage21.SI + stage25.SI + stage27.SI + stage35.SI)/6,
  somiteIIIvsII = (stage8.SIII + stage18.SIII + stage21.SIII + stage25.SIII + stage27.SIII + stage35.SIII)/6 - (stage8.SII + stage18.SII + stage21.SII + stage25.SII + stage27.SII + stage35.SII)/6,
  levels=design)

trios.average <- list()
trios.average[["somiteIIvsI"]] <- glmQLFTest(fit, contrast = my.contrasts[,"somiteIIvsI"])
trios.average[["somiteIIIvsI"]] <- glmQLFTest(fit, contrast = my.contrasts[,"somiteIIIvsI"])
trios.average[["somiteIIIvsII"]] <- glmQLFTest(fit, contrast = my.contrasts[,"somiteIIIvsII"])

## data.frames
somiteTrios.average.pca <- list()
for(contrast in paste0("somite",c("IIvsI","IIIvsI","IIIvsII"))){
  somiteTrios.average.pca[[contrast]] <- merged$region
  mcols(somiteTrios.average.pca[[contrast]]) <- combineTests(merged$id, trios.average[[contrast]]$table)
  best <- getBestTest(merged$id, trios.average[[contrast]]$table)
  mcols(somiteTrios.average.pca[[contrast]]) <- cbind(mcols(somiteTrios.average.pca[[contrast]]), best[,-c((ncol(best)-2):ncol(best))])
}
saveRDS(somiteTrios.average.pca, paste0(dir, "ATAC-seq/results/04.3_diffAccessibility_somiteTrios_average_pca.Rds"))

data.frame(somiteIvsII=sum(mcols(somiteTrios.average.pca[['somiteIIvsI']])$FDR <= 0.05),
           somiteIIvsIII=sum(mcols(somiteTrios.average.pca[['somiteIIIvsII']])$FDR <= 0.05),
           somiteIvsIII=sum(mcols(somiteTrios.average.pca[['somiteIIIvsI']])$FDR <= 0.05))
```

```
##   somiteIvsII somiteIIvsIII somiteIvsIII
## 1          84             3         1544
```

We only detect a large number of regions in the comparison between somites I and III, which are the most different. Again, these regions are likely to be monotonically increasing/decreasing as somites differentiate, and if changes are subtle they only become significant when comparing the extremes.


```r
s1 <- somiteTrios.average.pca[["somiteIIvsI"]][somiteTrios.average.pca[["somiteIIvsI"]]$FDR<0.05,]
s2 <- somiteTrios.average.pca[["somiteIIIvsI"]][somiteTrios.average.pca[["somiteIIIvsI"]]$FDR<0.05,]
s3 <- somiteTrios.average.pca[["somiteIIIvsII"]][somiteTrios.average.pca[["somiteIIIvsII"]]$FDR<0.05,]

# length(unique(queryHits(findOverlaps(s1, s3)))) # 0
# length(unique(queryHits(findOverlaps(s2, s3)))) # 3
```


```r
somiteTrios.average.pca.sig <- somiteTrios.average.pca[["somiteIIvsI"]][somiteTrios.average.pca[["somiteIIvsI"]]$FDR<0.05]
somiteTrios.average.pca.sig <- c(somiteTrios.average.pca.sig, somiteTrios.average.pca[["somiteIIIvsI"]][somiteTrios.average.pca[["somiteIIIvsI"]]$FDR<0.05])
somiteTrios.average.pca.sig <- c(somiteTrios.average.pca.sig, somiteTrios.average.pca[["somiteIIIvsII"]][somiteTrios.average.pca[["somiteIIIvsII"]]$FDR<0.05])
somiteTrios.average.pca.sig <- unique(somiteTrios.average.pca.sig)

somiteTrios.perStage.pca.sig <- somiteTrios.perStage.pca[["stage8"]][somiteTrios.perStage.pca[["stage8"]]$FDR<0.05]
somiteTrios.perStage.pca.sig <- c(somiteTrios.perStage.pca.sig, somiteTrios.perStage.pca[["stage18"]][somiteTrios.perStage.pca[["stage18"]]$FDR<0.05])
somiteTrios.perStage.pca.sig <- c(somiteTrios.perStage.pca.sig, somiteTrios.perStage.pca[["stage21"]][somiteTrios.perStage.pca[["stage21"]]$FDR<0.05])
somiteTrios.perStage.pca.sig <- c(somiteTrios.perStage.pca.sig, somiteTrios.perStage.pca[["stage25"]][somiteTrios.perStage.pca[["stage25"]]$FDR<0.05])
somiteTrios.perStage.pca.sig <- c(somiteTrios.perStage.pca.sig, somiteTrios.perStage.pca[["stage35"]][somiteTrios.perStage.pca[["stage35"]]$FDR<0.05])
somiteTrios.perStage.pca.sig <- unique(somiteTrios.perStage.pca.sig)
```

Overall, there are 1810 regions across the somites that reach significance on a per-stage basis, and 1581 when taking the average across stages. Interestingly, the majority of the per-stage regions are not recapitulated in the average test, with only a quarter significant in both comparisons.


```r
round(length(unique(queryHits(findOverlaps(somiteTrios.perStage.pca.sig, somiteTrios.average.pca.sig))))/length(somiteTrios.perStage.pca.sig)*100, 2)
```

```
## [1] 26.19
```

```r
# somiteTrios.perStage.pca.sig[somiteTrios.perStage.pca.sig %outside% somiteTrios.average.pca.sig] # 1336
```

That leaves 1,336 regions significant on a per-stage basis.

##### Differences across developmental stages

Now, we can also test for changes across development. First, we test all pairwise comparisons between stages on a per-somite basis. 


```r
my.contrasts <- makeContrasts(stage8vs18 = stage18.SI - stage8.SI, stage8vs21 = stage21.SI - stage8.SI,
                              stage8vs25 = stage25.SI - stage8.SI, stage8vs27 = stage27.SI - stage8.SI, 
                              stage8vs35 = stage35.SI - stage8.SI, 
                              stage18vs21 = stage21.SI - stage18.SI, stage18vs25 = stage25.SI - stage18.SI, 
                              stage18vs27 = stage27.SI - stage18.SI, stage18vs35 = stage35.SI - stage18.SI, 
                              stage21vs25 = stage25.SI - stage21.SI, stage21vs27 = stage27.SI - stage21.SI, 
                              stage21vs35 = stage35.SI - stage21.SI, 
                              stage25vs27 = stage27.SI - stage25.SI, stage25vs35 = stage35.SI - stage25.SI, 
                              stage27vs35 = stage35.SI - stage27.SI, levels=design)

development.perSomite <- list()
development.perSomite[["somiteI"]] <- glmQLFTest(fit, contrast = my.contrasts)

my.contrasts <- makeContrasts(stage8vs18 = stage18.SII - stage8.SII, stage8vs21 = stage21.SII - stage8.SII,
                              stage8vs25 = stage25.SII - stage8.SII, stage8vs27 = stage27.SII - stage8.SII, 
                              stage8vs35 = stage35.SII - stage8.SII, 
                              stage18vs21 = stage21.SII - stage18.SII, stage18vs25 = stage25.SII - stage18.SII, 
                              stage18vs27 = stage27.SII - stage18.SII, stage18vs35 = stage35.SII - stage18.SII, 
                              stage21vs25 = stage25.SII - stage21.SII, stage21vs27 = stage27.SII - stage21.SII, 
                              stage21vs35 = stage35.SII - stage21.SII, 
                              stage25vs27 = stage27.SII - stage25.SII, stage25vs35 = stage35.SII - stage25.SII, 
                              stage27vs35 = stage35.SII - stage27.SII, levels=design)
development.perSomite[["somiteII"]] <- glmQLFTest(fit, contrast = my.contrasts)

my.contrasts <- makeContrasts(stage8vs18 = stage18.SIII - stage8.SIII, stage8vs21 = stage21.SIII - stage8.SIII,
                              stage8vs25 = stage25.SIII - stage8.SIII, stage8vs27 = stage27.SIII - stage8.SIII, 
                              stage8vs35 = stage35.SIII - stage8.SIII, 
                              stage18vs21 = stage21.SIII - stage18.SIII, stage18vs25 = stage25.SIII - stage18.SIII, 
                              stage18vs27 = stage27.SIII - stage18.SIII, stage18vs35 = stage35.SIII - stage18.SIII, 
                              stage21vs25 = stage25.SIII - stage21.SIII, stage21vs27 = stage27.SIII - stage21.SIII, 
                              stage21vs35 = stage35.SIII - stage21.SIII, 
                              stage25vs27 = stage27.SIII - stage25.SIII, stage25vs35 = stage35.SIII - stage25.SIII, 
                              stage27vs35 = stage35.SIII - stage27.SIII, levels=design)
development.perSomite[["somiteIII"]] <- glmQLFTest(fit, contrast = my.contrasts)
```

This gives us a few thousand significant regions, with many more recovered for somiteI.


```r
stage.perSomite.pca <- list()
for(contrast in paste0("somite",c("I","II","III"))){
  stage.perSomite.pca[[contrast]] <- merged$region
  mcols(stage.perSomite.pca[[contrast]]) <- combineTests(merged$id, development.perSomite[[contrast]]$table)
  best <- getBestTest(merged$id, development.perSomite[[contrast]]$table)
  mcols(stage.perSomite.pca[[contrast]]) <- cbind(mcols(stage.perSomite.pca[[contrast]]), best[,-c((ncol(best)-2):ncol(best))])
}
saveRDS(stage.perSomite.pca, paste0(dir, "ATAC-seq/results/04.3_diffAccessibility_stages_perSomite_pca.Rds"))

df <- t(data.frame(somiteI=sum(mcols(stage.perSomite.pca[['somiteI']])$FDR <= 0.05),
           somiteII=sum(mcols(stage.perSomite.pca[['somiteII']])$FDR <= 0.05), 
           somiteIII=sum(mcols(stage.perSomite.pca[['somiteIII']])$FDR <= 0.05)))
colnames(df) <- "nDA"
df
```

```
##             nDA
## somiteI   14207
## somiteII   3900
## somiteIII  4402
```

There are large numbers of regions only detected in a specific somite. These probably contain some modest changes that are not significant for the other somites, but likely also some somite-specific changes.


```r
p <- makeVennDiagram(list(somiteI=stage.perSomite.pca[["somiteI"]][stage.perSomite.pca[["somiteI"]]$FDR < 0.05,],
                          somiteII=stage.perSomite.pca[["somiteII"]][stage.perSomite.pca[["somiteII"]]$FDR < 0.05,],
                          somiteIII=stage.perSomite.pca[["somiteIII"]][stage.perSomite.pca[["somiteIII"]]$FDR < 0.05,]))
```

![](04.3_differentialAccessibility_PCA_files/figure-html/overlap_stage_somite-1.png)<!-- -->


Now, we instead take the average of somites I, II and III. This should increase power.


```r
my.contrasts <- makeContrasts(stage8vs18 = (stage18.SIII+stage18.SII+stage18.SI)/3 - (stage8.SIII+stage8.SII+stage8.SI)/3,
                              stage8vs21 = (stage21.SIII+stage21.SII+stage21.SI)/3 - (stage8.SIII+stage8.SII+stage8.SI)/3,
                              stage8vs25 = (stage25.SIII+stage25.SII+stage25.SI)/3 - (stage8.SIII+stage8.SII+stage8.SI)/3,
                              stage8vs27 = (stage27.SIII+stage27.SII+stage27.SI)/3 - (stage8.SIII+stage8.SII+stage8.SI)/3,
                              stage8vs35 = (stage35.SIII+stage35.SII+stage35.SI)/3 - (stage8.SIII+stage8.SII+stage8.SI)/3,
                              stage18vs21 = (stage21.SIII+stage21.SII+stage21.SI)/3 - (stage18.SIII+stage18.SII+stage18.SI)/3,
                              stage18vs25 = (stage25.SIII+stage25.SII+stage25.SI)/3 - (stage18.SIII+stage18.SII+stage18.SI)/3,
                              stage18vs27 = (stage27.SIII+stage27.SII+stage27.SI)/3 - (stage18.SIII+stage18.SII+stage18.SI)/3,
                              stage18vs35 = (stage35.SIII+stage35.SII+stage35.SI)/3 - (stage18.SIII+stage18.SII+stage18.SI)/3,
                              stage21vs25 = (stage25.SIII+stage25.SII+stage25.SI)/3 - (stage21.SIII+stage21.SII+stage21.SI)/3,
                              stage21vs27 = (stage27.SIII+stage27.SII+stage27.SI)/3 - (stage21.SIII+stage21.SII+stage21.SI)/3,
                              stage21vs35 = (stage35.SIII+stage35.SII+stage35.SI)/3 - (stage21.SIII+stage21.SII+stage21.SI)/3,
                              stage25vs27 = (stage27.SIII+stage27.SII+stage27.SI)/3 - (stage25.SIII+stage25.SII+stage25.SI)/3,
                              stage25vs35 = (stage35.SIII+stage35.SII+stage35.SI)/3 - (stage25.SIII+stage25.SII+stage25.SI)/3,
                              stage27vs35 = (stage35.SIII+stage35.SII+stage35.SI)/3 - (stage27.SIII+stage27.SII+stage27.SI)/3, levels=design)

development.average <- glmQLFTest(fit, contrast = my.contrasts)
```

And we test against the null that none of the stages are different. Therefore, anything with an FDR of 5% or lower indicates a region of open chromatin that is differentially accessible between at least a pair of stages. 

Nearly 30 thousand regions are differentially accessible across development.


```r
stage.average.pca <- merged$region
mcols(stage.average.pca) <- combineTests(merged$id, development.average$table)
best <- getBestTest(merged$id, development.average$table)
mcols(stage.average.pca) <- cbind(mcols(stage.average.pca), best[,-c((ncol(best)-2):ncol(best))])

summary(stage.average.pca$FDR<0.05)
```

```
##    Mode   FALSE    TRUE 
## logical  145593   29648
```

```r
saveRDS(stage.average.pca, paste0(dir, "ATAC-seq/results/04.3_diffAccessibility_stages_average_pca.Rds"))
```

Most of the regions identified on a per-somite basis are also significant when taking their average.


```r
stage.average.pca.sig <- stage.average.pca[stage.average.pca$FDR<0.05]
stage.perSomite.pca.sig <- stage.perSomite.pca[["somiteI"]][stage.perSomite.pca[["somiteI"]]$FDR<0.05]
stage.perSomite.pca.sig <- c(stage.perSomite.pca.sig, stage.perSomite.pca[["somiteII"]][stage.perSomite.pca[["somiteII"]]$FDR<0.05])
stage.perSomite.pca.sig <- c(stage.perSomite.pca.sig, stage.perSomite.pca[["somiteIII"]][stage.perSomite.pca[["somiteIII"]]$FDR<0.05])
stage.perSomite.pca.sig <- unique(stage.perSomite.pca.sig)

round(length(unique(queryHits(findOverlaps(stage.perSomite.pca.sig, stage.average.pca.sig))))/length(stage.perSomite.pca.sig)*100, 2)
```

```
## [1] 76.54
```

```r
# stage.perSomite.pca.sig[stage.perSomite.pca.sig %outside% stage.average.pca.sig] # 4090
```

Nonetheless, 4,090 regions are only significant on a per-somite basis.



```r
sessionInfo()
```

```
## R version 3.5.1 (2018-07-02)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS: /datastore/sw/gentoo2/usr/lib64/blas/reference/libblas.so.3.7.0
## LAPACK: /datastore/sw/gentoo2/usr/lib64/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
##  [1] grid      parallel  stats4    stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] ChIPpeakAnno_3.16.1         VennDiagram_1.6.20         
##  [3] futile.logger_1.4.3         Biostrings_2.50.2          
##  [5] XVector_0.22.0              edgeR_3.24.3               
##  [7] limma_3.38.3                csaw_1.16.1                
##  [9] SummarizedExperiment_1.12.0 DelayedArray_0.8.0         
## [11] BiocParallel_1.16.6         matrixStats_0.54.0         
## [13] Biobase_2.42.0              GenomicRanges_1.34.0       
## [15] GenomeInfoDb_1.18.2         IRanges_2.16.0             
## [17] S4Vectors_0.20.1            BiocGenerics_0.28.0        
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.4.0               idr_1.2                 
##  [3] regioneR_1.14.0          bit64_0.9-7             
##  [5] splines_3.5.1            assertthat_0.2.1        
##  [7] statmod_1.4.30           BiocManager_1.30.4      
##  [9] RBGL_1.58.2              blob_1.1.1              
## [11] BSgenome_1.50.0          GenomeInfoDbData_1.2.0  
## [13] Rsamtools_1.34.1         yaml_2.2.0              
## [15] progress_1.2.0           RSQLite_2.1.1           
## [17] lattice_0.20-35          digest_0.6.18           
## [19] htmltools_0.3.6          Matrix_1.2-14           
## [21] XML_3.98-1.15            pkgconfig_2.0.2         
## [23] biomaRt_2.38.0           zlibbioc_1.28.0         
## [25] GO.db_3.7.0              AnnotationFilter_1.6.0  
## [27] GenomicFeatures_1.34.8   lazyeval_0.2.2          
## [29] survival_2.42-6          magrittr_1.5            
## [31] crayon_1.3.4             memoise_1.1.0           
## [33] evaluate_0.13            MASS_7.3-50             
## [35] graph_1.60.0             tools_3.5.1             
## [37] prettyunits_1.0.2        hms_0.4.2               
## [39] formatR_1.6              stringr_1.4.0           
## [41] locfit_1.5-9.1           ensembldb_2.6.8         
## [43] AnnotationDbi_1.44.0     lambda.r_1.2.3          
## [45] ade4_1.7-13              compiler_3.5.1          
## [47] rlang_0.3.4              RCurl_1.95-4.12         
## [49] bitops_1.0-6             rmarkdown_1.12          
## [51] multtest_2.38.0          curl_3.3                
## [53] DBI_1.0.0                R6_2.4.0                
## [55] GenomicAlignments_1.18.1 knitr_1.22              
## [57] rtracklayer_1.42.2       seqinr_3.4-5            
## [59] bit_1.1-14               ProtGenerics_1.14.0     
## [61] futile.options_1.0.1     stringi_1.4.3           
## [63] Rcpp_1.0.1               xfun_0.6
```

