---
title: "Overlaps of DA regions"
date: "15 April 2019"
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---



We have performed differential accessibility (DA) analysis on the ATAC-seq data, testing for differences between the somite trios and between the different stages. To effectively remove technical biases, we computed the residuals from the fit of the `somite*stage` interaction, and performed PCA on them; we used the first 18 PCs as covariates in the DA analysis (18 PCs were chosen based on parallel analysis, which uses permutations of the count data to calculate the amount of variance explained under a null model; only PCs that explain more variance than under the null are retained). The number of significant regions are summarised below:


```r
## metadata
meta <- read.table(paste0(dir, "ATAC-seq/data/metadata_ATACseq.tsv"), stringsAsFactors = FALSE, header = TRUE)
meta <- meta[meta$QCpass==1,]

### DA regions
### merge the results from all different comparisons into a single table, keeping the FDR for each comparison for each region

## somite trios
# per stage
trios.perStage <- readRDS(paste0(dir, "ATAC-seq/results/04.3_diffAccessibility_somiteTrios_perStage_pca.Rds"))
for(i in 1:length(trios.perStage)){
  trios.perStage[[i]] <- trios.perStage[[i]][order(trios.perStage[[i]])]
}
stopifnot(identical(ranges(trios.perStage[["stage8"]]), ranges(trios.perStage[["stage18"]])))
stopifnot(identical(ranges(trios.perStage[["stage21"]]), ranges(trios.perStage[["stage35"]])))

trios.perStage.all <- as.data.frame(trios.perStage[["stage8"]])[,c(1:4,6,14)]
colnames(trios.perStage.all)[ncol(trios.perStage.all)] <- "stage8"
trios.perStage.all$stage18 <- trios.perStage[["stage18"]]$FDR
trios.perStage.all$stage21 <- trios.perStage[["stage21"]]$FDR
trios.perStage.all$stage25 <- trios.perStage[["stage25"]]$FDR
trios.perStage.all$stage35 <- trios.perStage[["stage35"]]$FDR
trios.perStage.all <- subset(trios.perStage.all, trios.perStage.all$stage8 < 0.05 | trios.perStage.all$stage18 < 0.05 | trios.perStage.all$stage21 < 0.05 | trios.perStage.all$stage25 < 0.05 | trios.perStage.all$stage35 < 0.05)
row.names(trios.perStage.all) <- paste0(trios.perStage.all$seqnames, ":", trios.perStage.all$start, "-", trios.perStage.all$end)

nDE <- t(data.frame(stage8 = nrow(trios.perStage.all[trios.perStage.all$stage8 < 0.05,]),
                  stage18 = nrow(trios.perStage.all[trios.perStage.all$stage18 < 0.05,]),
                  stage21 = nrow(trios.perStage.all[trios.perStage.all$stage21 < 0.05,]),
                  stage25 = nrow(trios.perStage.all[trios.perStage.all$stage25 < 0.05,]),
                  stage35 = nrow(trios.perStage.all[trios.perStage.all$stage35 < 0.05,])))
colnames(nDE) <- "trios.perStage"
nDE
```

```
##         trios.perStage
## stage8             573
## stage18            193
## stage21              6
## stage25           1094
## stage35             46
```

```r
# stage average
trios.average <- readRDS(paste0(dir, "ATAC-seq/results/04.3_diffAccessibility_somiteTrios_average_pca.Rds"))
for(i in 1:length(trios.average)){
  trios.average[[i]] <- trios.average[[i]][order(trios.average[[i]])]
}
stopifnot(identical(ranges(trios.average[["somiteIIvsI"]]), ranges(trios.average[["somiteIIIvsI"]])))
stopifnot(identical(ranges(trios.average[["somiteIIIvsI"]]), ranges(trios.average[["somiteIIIvsII"]])))

trios.average.all <- as.data.frame(trios.average[["somiteIIvsI"]])[,c(1:4,6:8,10)]
colnames(trios.average.all)[ncol(trios.average.all)] <- "somiteIIvsI"
trios.average.all$somiteIIIvsII <- trios.average[["somiteIIIvsII"]]$FDR
trios.average.all$somiteIIIvsI <- trios.average[["somiteIIIvsI"]]$FDR
trios.average.all <- subset(trios.average.all, trios.average.all$somiteIIvsI < 0.05 | trios.average.all$somiteIIIvsII < 0.05 | trios.average.all$somiteIIIvsI < 0.05)
row.names(trios.average.all) <- paste0(trios.average.all$seqnames, ":", trios.average.all$start, "-", trios.average.all$end)

nDE <- t(data.frame(somiteIvsII = nrow(trios.average.all[trios.average.all$somiteIIvsI < 0.05,]),
                  somiteIIvsIII = nrow(trios.average.all[trios.average.all$somiteIIIvsII < 0.05,]),
                  somiteIvsIII = nrow(trios.average.all[trios.average.all$somiteIIIvsI < 0.05,])))
colnames(nDE) <- "trios.average"
nDE
```

```
##               trios.average
## somiteIvsII              84
## somiteIIvsIII             3
## somiteIvsIII           1544
```

```r
## stages
# per somite
stage.perSomite <- readRDS(paste0(dir, "ATAC-seq/results/04.3_diffAccessibility_stages_perSomite_pca.Rds"))
for(i in 1:length(stage.perSomite)){
  stage.perSomite[[i]] <- stage.perSomite[[i]][order(stage.perSomite[[i]])]
}
stopifnot(identical(ranges(stage.perSomite[["somiteI"]]), ranges(stage.perSomite[["somiteII"]])))
stopifnot(identical(ranges(stage.perSomite[["somiteI"]]), ranges(stage.perSomite[["somiteIII"]])))

stage.perSomite.all <- as.data.frame(stage.perSomite[["somiteI"]])[,c(1:4,6,36)]
colnames(stage.perSomite.all)[ncol(stage.perSomite.all)] <- "somiteI"
stage.perSomite.all$somiteII <- stage.perSomite[["somiteII"]]$FDR
stage.perSomite.all$somiteIII <- stage.perSomite[["somiteIII"]]$FDR
stage.perSomite.all <- subset(stage.perSomite.all, stage.perSomite.all$somiteI < 0.05 | stage.perSomite.all$somiteII < 0.05 | stage.perSomite.all$somiteIII < 0.05)
row.names(stage.perSomite.all) <- paste0(stage.perSomite.all$seqnames, ":", stage.perSomite.all$start, "-", stage.perSomite.all$end)

nDE <- t(data.frame(somiteI = nrow(stage.perSomite.all[stage.perSomite.all$somiteI < 0.05,]),
                    somiteII = nrow(stage.perSomite.all[stage.perSomite.all$somiteII < 0.05,]),
                    somiteIII = nrow(stage.perSomite.all[stage.perSomite.all$somiteIII < 0.05,])))
colnames(nDE) <- "stages.perSomite"
nDE
```

```
##           stages.perSomite
## somiteI              14207
## somiteII              3900
## somiteIII             4402
```

```r
# somite average
stage.average.all <- as.data.frame(readRDS(paste0(dir, "ATAC-seq/results/04.3_diffAccessibility_stages_average_pca.Rds")))[,c(1:4,6,38)]
stage.average.all <- stage.average.all[stage.average.all$FDR < 0.05,]
row.names(stage.average.all) <- paste0(stage.average.all$seqnames, ":", stage.average.all$start, "-", stage.average.all$end)

nDE <- t(data.frame(any = nrow(stage.average.all)))
colnames(nDE) <- "stages.average"
nDE
```

```
##     stages.average
## any          29648
```

### Overlap between average and somite/stage-specific tests

The analyses with the average profiles across stages (for somite trios comparisons) or across somites (for stage comparisons) have the most statistical power and thus detect the largest number of significant regions. But any stage- or somite-specific change are lost.

Interestingly, the overlap between the average versus stage-specific results for the somite trios comparisons is small, with only a quarter of the regions recovered in the stage-specific tests also significant when the average is taken.


```r
ave <- GRanges(trios.average.all$seqnames, IRanges(trios.average.all$start, trios.average.all$end))
perStage <- GRanges(trios.perStage.all$seqnames, IRanges(trios.perStage.all$start, trios.perStage.all$end), stage8=trios.perStage.all$stage8, stage18=trios.perStage.all$stage18, stage21=trios.perStage.all$stage21, stage25=trios.perStage.all$stage25, stage35=trios.perStage.all$stage35)

nDE <- matrix(c(length(perStage), length(ave), length(unique(subjectHits(findOverlaps(ave, perStage)))), length(unique(queryHits(findOverlaps(ave, perStage))))), ncol=2)
colnames(nDE) <- c("total", "overlap")
row.names(nDE) <- c("perStage", "average")
nDE
```

```
##          total overlap
## perStage  1810     474
## average   1581     469
```

Also, most of the differences identified within each stage are not shared with the other stages.


```r
p <- makeVennDiagram(list(stage8 = trios.perStage[["stage8"]][trios.perStage[["stage8"]]$FDR<0.05,],
                     stage18 = trios.perStage[["stage18"]][trios.perStage[["stage18"]]$FDR<0.05,],
                     stage21 = trios.perStage[["stage21"]][trios.perStage[["stage21"]]$FDR<0.05,],
                     stage25 = trios.perStage[["stage25"]][trios.perStage[["stage25"]]$FDR<0.05,],
                     stage35 = trios.perStage[["stage35"]][trios.perStage[["stage35"]]$FDR<0.05,]), 
                     fill=c("tomato1", "darkolivegreen4", "steelblue", "darkorange1", "purple2"), alpha=0.2,
                     cat.col=c("tomato1", "darkolivegreen4", "steelblue", "darkorange1", "purple2"),
                     cat.fontface=2, cat.cex=0.9)
```

![](04.5_comparison_DAregions_stages_somiteTrios_files/figure-html/trios.perstage-1.png)<!-- -->

There is a large number of regions identified when only using the stage25 data, and most are not recovered with the average test, so it'll be interesting to see what these regions are.

In terms of the stage comparisons, most of the differences on a per-somite basis are also significant when taking their average.


```r
ave <- GRanges(stage.average.all$seqnames, IRanges(stage.average.all$start, stage.average.all$end))
perSomite <- GRanges(stage.perSomite.all$seqnames, IRanges(stage.perSomite.all$start, stage.perSomite.all$end))

nDE <- matrix(c(length(perSomite), length(ave), length(unique(subjectHits(findOverlaps(ave, perSomite)))), length(unique(queryHits(findOverlaps(ave, perSomite))))), ncol=2)
colnames(nDE) <- c("total", "overlap")
row.names(nDE) <- c("perSomite", "average")
nDE
```

```
##           total overlap
## perSomite 17434   13344
## average   29648   14136
```

In terms of the per-somite results, somiteI data returns a much higher number of differences; 14,207 significant regions of which 75% are also recovered in the average analysis This suggests that the difference in numbers comes from increased power (we have several replicates for all stages for somite I but not for the others) but also some of these regions might be significant in the average test driven by the somiteI behaviour. 

For somites II and III, most of the regions are shared with the other somites, with about ~30% of their significant regions private.


```r
p <- makeVennDiagram(list(somiteI = stage.perSomite[["somiteI"]][stage.perSomite[["somiteI"]]$FDR<0.05,],
                          somiteII = stage.perSomite[["somiteII"]][stage.perSomite[["somiteII"]]$FDR<0.05,],
                          somiteIII = stage.perSomite[["somiteIII"]][stage.perSomite[["somiteIII"]]$FDR<0.05,]),
                     fill=c("tomato1", "darkolivegreen4", "steelblue"), alpha=0.2,
                     cat.col=c("tomato1", "darkolivegreen4", "steelblue"),
                     cat.fontface=2, cat.cex=0.9)
```

![](04.5_comparison_DAregions_stages_somiteTrios_files/figure-html/stages.persomite-1.png)<!-- -->

### Peak sets

For downstream analyses, we create independent sets of DA regions:

- **Robust set**: regions identified in average test plus at least one stage/somite-specific test.
- **Power set**: regions identified only in average test.
- **Specific set**: regions identified only in stage/somite-specific test; split by the comparison of relevance.


```r
## somite trios
trios.robust <- trios.average.all[intersect(row.names(trios.average.all), row.names(trios.perStage.all)),] # 439
trios.power <- trios.average.all[setdiff(row.names(trios.average.all), row.names(trios.robust)),] # 1,142
trios.specific <- trios.perStage.all[setdiff(row.names(trios.perStage.all), row.names(trios.robust)),] # 1,371
trios.specific.stage8 <- trios.specific[trios.specific$stage8 < 0.05 & 
                                          trios.specific$stage18 > 0.05 & 
                                          trios.specific$stage21 > 0.05 & 
                                          trios.specific$stage25 > 0.05 & 
                                          trios.specific$stage35 > 0.05,] # 462
trios.specific.stage18 <- trios.specific[trios.specific$stage18 < 0.05 & 
                                          trios.specific$stage8 > 0.05 & 
                                           trios.specific$stage21 > 0.05 & 
                                          trios.specific$stage25 > 0.05 & 
                                           trios.specific$stage35 > 0.05,] # 94
trios.specific.stage21 <- trios.specific[trios.specific$stage21 < 0.05 & 
                                          trios.specific$stage8 > 0.05 & 
                                           trios.specific$stage18 > 0.05 & 
                                          trios.specific$stage25 > 0.05 & 
                                           trios.specific$stage35 > 0.05,] # 2
trios.specific.stage25 <- trios.specific[trios.specific$stage25 < 0.05 & 
                                          trios.specific$stage8 > 0.05 & 
                                           trios.specific$stage18 > 0.05 & 
                                          trios.specific$stage21 > 0.05 & 
                                           trios.specific$stage35 > 0.05,] # 755
trios.specific.stage35 <- trios.specific[trios.specific$stage35 < 0.05 & 
                                          trios.specific$stage8 > 0.05 & 
                                           trios.specific$stage18 > 0.05 & 
                                          trios.specific$stage21 > 0.05 & 
                                           trios.specific$stage25 > 0.05,] # 31
# 27 regions significant in two or more stages

## stage changes
stages.robust <- stage.average.all[intersect(row.names(stage.average.all), row.names(stage.perSomite.all)),] # 12,941
stages.power <- stage.average.all[setdiff(row.names(stage.average.all), row.names(stages.robust)),] # 16,707
stages.specific <- stage.perSomite.all[setdiff(row.names(stage.perSomite.all), row.names(stages.robust)),] # 4,493
stages.specific.somiteI <- stages.specific[stages.specific$somiteI < 0.05 &
                                             stages.specific$somiteII > 0.05 & 
                                             stages.specific$somiteIII > 0.05,] # 3,884
stages.specific.somiteII <- stages.specific[stages.specific$somiteII < 0.05 &
                                              stages.specific$somiteI > 0.05 & 
                                              stages.specific$somiteIII > 0.05,] # 283
stages.specific.somiteIII <- stages.specific[stages.specific$somiteIII < 0.05 &
                                               stages.specific$somiteI > 0.05 & 
                                               stages.specific$somiteI > 0.05,] # 290
# 36 regions significant in two comparisons

save(trios.robust, trios.power, trios.specific, trios.specific.stage8, trios.specific.stage18, trios.specific.stage21, trios.specific.stage25, trios.specific.stage35, stages.robust, stages.power, stages.specific, stages.specific.somiteI, stages.specific.somiteII, stages.specific.somiteIII, file=paste0(dir, "ATAC-seq/results/04.5_DApeakSets.RData"))
```

### Genomic features of DA regions

Most open chromatin regions overlap or are near genes, but only about a fifth are in promoters. Thus, the vast majority of regions are putative regulatory elements. Also, DA regions are slightly enriched for more intergenic, compared to genic regions.


```r
peakCounts <- readRDS(paste0(dir, "ATAC-seq/results/04.4_peakCounts_csawMerged.Rds"))
peakSet <- rowRanges(peakCounts)
ann <- detailRanges(peakSet, TxDb.Mmusculus.UCSC.mm10.knownGene, org.Mm.eg.db, dist=5000, promoter = c(3000,300))
mcols(peakSet) <- ann

peakSet$DAtrios <- 0
peakSet$DAstages <- 0

tmp <- GRanges(trios.average.all$seqnames, IRanges(trios.average.all$start, trios.average.all$end))
tmp <- c(tmp, GRanges(trios.perStage.all$seqnames, IRanges(trios.perStage.all$start, trios.perStage.all$end)))
tmp <- unique(tmp)
peakSet[peakSet %in% tmp]$DAtrios <- 1

tmp <- GRanges(stage.average.all$seqnames, IRanges(stage.average.all$start, stage.average.all$end))
tmp <- c(tmp, GRanges(stage.perSomite.all$seqnames, IRanges(stage.perSomite.all$start, stage.perSomite.all$end)))
tmp <- unique(tmp)
peakSet[peakSet %in% tmp]$DAstages <- 1
saveRDS(peakSet, file=paste0(dir, "ATAC-seq/results/04.5_peakSet_csawMerged_geneAnn.Rds"))

# table(trios=peakSet$DAtrios, stages=peakSet$DAstages)
#      stages
# trios      0      1
#     0 140295  31994
#     1    805   2147

nDE <- matrix(c(
  length(peakSet[peakSet$overlap != ""])/length(peakSet)*100,
  length(peakSet[peakSet$overlap != "" & peakSet$DAtrios == 1])/length(peakSet[peakSet$DAtrios == 1])*100,
  length(peakSet[peakSet$overlap != "" & peakSet$DAstages == 1])/length(peakSet[peakSet$DAstages == 1])*100,
  
  length(peakSet[grep(":P", peakSet$overlap)])/length(peakSet)*100,
  length(peakSet[peakSet$DAtrios==1][grep(":P", peakSet[peakSet$DAtrios==1]$overlap)])/length(peakSet[peakSet$DAtrios==1])*100,
  length(peakSet[peakSet$DAstages==1][grep(":P", peakSet[peakSet$DAstages==1]$overlap)])/length(peakSet[peakSet$DAstages==1])*100,
  
  length(peakSet[peakSet$overlap=="" & (peakSet$left != "" | peakSet$right != "")])/length(peakSet)*100,
  length(peakSet[peakSet$DAtrios==1 & peakSet$overlap=="" & (peakSet$left != "" | peakSet$right != "")])/length(peakSet[peakSet$DAtrios == 1])*100,
  length(peakSet[peakSet$DAstages==1 & peakSet$overlap=="" & (peakSet$left != "" | peakSet$right != "")])/length(peakSet[peakSet$DAstages == 1])*100,
  
  length(peakSet[peakSet$overlap=="" & peakSet$left == "" & peakSet$right==""])/length(peakSet)*100,
  length(peakSet[peakSet$DAtrios==1 & peakSet$overlap=="" & peakSet$left == "" & peakSet$right==""])/length(peakSet[peakSet$DAtrios==1])*100,
  length(peakSet[peakSet$DAstages==1 & peakSet$overlap=="" & peakSet$left == "" & peakSet$right==""])/length(peakSet[peakSet$DAstages==1])*100
), ncol=3, byrow = TRUE)
colnames(nDE) <- c("all", "DAtrios", "DAstages")
row.names(nDE) <- c("overlapGenes", "inPromoter", "nearGenes", "intergenic")
round(nDE,2)
```

```
##                all DAtrios DAstages
## overlapGenes 63.39   52.88    59.80
## inPromoter   18.60    9.59    16.60
## nearGenes     5.89    4.98     4.40
## intergenic   30.72   42.14    35.81
```

```r
# table(unlist(lapply(strsplit(unlist(strsplit(peakSet[peakSet$overlap != ""]$overlap, ",")), ":"), "[[", 3)))
#     E     I     P    PE    PI 
# 20359 66775 10323 23955  2471
```





```r
sessionInfo()
```

```
## R version 3.5.1 (2018-07-02)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: CentOS Linux 7 (Core)
## 
## Matrix products: default
## BLAS: /datastore/sw/gentoo2/usr/lib64/blas/reference/libblas.so.3.7.0
## LAPACK: /datastore/sw/gentoo2/usr/lib64/R/lib/libRlapack.so
## 
## locale:
##  [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    
##  [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8   
##  [7] LC_PAPER=en_GB.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
##  [1] grid      parallel  stats4    stats     graphics  grDevices utils    
##  [8] datasets  methods   base     
## 
## other attached packages:
##  [1] org.Mm.eg.db_3.7.0                      
##  [2] TxDb.Mmusculus.UCSC.mm10.knownGene_3.4.4
##  [3] GenomicFeatures_1.34.8                  
##  [4] AnnotationDbi_1.44.0                    
##  [5] ChIPpeakAnno_3.16.1                     
##  [6] VennDiagram_1.6.20                      
##  [7] futile.logger_1.4.3                     
##  [8] Biostrings_2.50.2                       
##  [9] XVector_0.22.0                          
## [10] csaw_1.16.1                             
## [11] SummarizedExperiment_1.12.0             
## [12] DelayedArray_0.8.0                      
## [13] BiocParallel_1.16.6                     
## [14] matrixStats_0.54.0                      
## [15] Biobase_2.42.0                          
## [16] GenomicRanges_1.34.0                    
## [17] GenomeInfoDb_1.18.2                     
## [18] IRanges_2.16.0                          
## [19] S4Vectors_0.20.1                        
## [20] BiocGenerics_0.28.0                     
## 
## loaded via a namespace (and not attached):
##  [1] httr_1.4.0               idr_1.2                 
##  [3] edgeR_3.24.3             regioneR_1.14.0         
##  [5] bit64_0.9-7              splines_3.5.1           
##  [7] assertthat_0.2.1         BiocManager_1.30.4      
##  [9] RBGL_1.58.2              blob_1.1.1              
## [11] BSgenome_1.50.0          GenomeInfoDbData_1.2.0  
## [13] Rsamtools_1.34.1         yaml_2.2.0              
## [15] progress_1.2.0           RSQLite_2.1.1           
## [17] lattice_0.20-35          limma_3.38.3            
## [19] digest_0.6.18            htmltools_0.3.6         
## [21] Matrix_1.2-14            XML_3.98-1.15           
## [23] pkgconfig_2.0.2          biomaRt_2.38.0          
## [25] zlibbioc_1.28.0          GO.db_3.7.0             
## [27] AnnotationFilter_1.6.0   lazyeval_0.2.2          
## [29] survival_2.42-6          magrittr_1.5            
## [31] crayon_1.3.4             memoise_1.1.0           
## [33] evaluate_0.13            MASS_7.3-50             
## [35] graph_1.60.0             tools_3.5.1             
## [37] prettyunits_1.0.2        hms_0.4.2               
## [39] formatR_1.6              stringr_1.4.0           
## [41] locfit_1.5-9.1           ensembldb_2.6.8         
## [43] lambda.r_1.2.3           ade4_1.7-13             
## [45] compiler_3.5.1           rlang_0.3.4             
## [47] RCurl_1.95-4.12          bitops_1.0-6            
## [49] rmarkdown_1.12           multtest_2.38.0         
## [51] curl_3.3                 DBI_1.0.0               
## [53] R6_2.4.0                 GenomicAlignments_1.18.1
## [55] knitr_1.22               rtracklayer_1.42.2      
## [57] seqinr_3.4-5             bit_1.1-14              
## [59] ProtGenerics_1.14.0      futile.options_1.0.1    
## [61] stringi_1.4.3            Rcpp_1.0.1              
## [63] xfun_0.6
```

