---
title: "Transcription factor binding sites"
date: "15 April 2019"
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(csaw)
library(ChIPpeakAnno)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(ggplot2)
library(ggrepel)
library(gplots)
library(RColorBrewer)
library(BSgenome.Mmusculus.UCSC.mm10)

dir <- "/user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/"

#### FUNCTIONS
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)
  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)
  numPlots = length(plots)
  
  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                     ncol = cols, nrow = ceiling(numPlots/cols), byrow = TRUE)
  }
  
  if (numPlots==1) {
    print(plots[[1]])
    
  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
    
    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
      
      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}
```

We have processed and analysed the RNA-seq and ATAC-seq data from mouse somites. We have gene expression and peak abundance levels, plus lists of differentially expressed/accessible genes and regions. 

```{r meatadata}
## metadata
meta <- read.table(paste0(dir, "ATAC-seq/data/metadata_ATACseq.tsv"), stringsAsFactors = FALSE, header = TRUE)
meta <- meta[meta$QCpass==1,]

## peaks
peakSet <- readRDS(file=paste0(dir, "ATAC-seq/results/04.5_peakSet_csawMerged_geneAnn.Rds"))

## gene expression
rnaseq <- read.table(paste0(dir, "RNA-seq/data/normalisedCounts_batchCorrected.tab"))
rnaseq$mean <- apply(rnaseq[,-1], 1, mean)

gene.ann <- read.table(paste0(dir, "RNA-seq/data/Mus_musculus.GRCm38.93.ann"))
gene.ann <- GRanges(paste0("chr", gene.ann$V3), IRanges(gene.ann$V4, gene.ann$V5), strand = gene.ann$V6, gene=gene.ann$V2, ID=gene.ann$V1)
```

### Transcription-factor binding analysis 

Now we want to identify interesting candidates important in the regulation of somitogenesis. One approach is to perform a transcription-factor centric analysis, where we focus on elements putatively regulated by particular transcription factors.

```{r readTFmatches_mm10}
## HOCOMOCO v10 PWMs used to scan the mm10 genome with PWMscan, default parameters.
## Taken from diffTF [https://www.embl.de/download/zaugg/diffTF/TFBS/TFBS_mm10_PWMScan_HOCOMOCOv10.tar.gz]
## removed ENOA which has been retracted in HOCOMOCOv11
tfbs <- list()
files <- list.files(paste0(dir, "ATAC-seq/motifs/PWMScan_HOCOMOCOv10_mm10/"))
for(file in files){
  tf <- substr(file, 1, nchar(file)-9)
  tfbs[[tf]] <- read.table(paste0(dir, "ATAC-seq/motifs/PWMScan_HOCOMOCOv10_mm10/", file))
  tfbs[[tf]] <- GRanges(tfbs[[tf]]$V1, IRanges(tfbs[[tf]]$V2, tfbs[[tf]]$V3), strand=tfbs[[tf]]$V6, motif=tfbs[[tf]]$V4, score=tfbs[[tf]]$V5)
}
saveRDS(tfbs, paste0(dir, "ATAC-seq/motifs/PWMScan_HOCOMOCOv10_mm10.Rds"))
# tfbs <- readRDS(paste0(dir, "ATAC-seq/motifs/PWMScan_HOCOMOCOv10_mm10.Rds"))
```

We will use the PWMs from HOCOMOCO, a project based on ChIP-seq data that should provide high-quality motifs. Each PWM is used to scan the mouse refence genome and predict TF binding sites. We read the coordinates of such putative TFBSs for all `r length(tfbs)` TFs available.

Most TFs have around 100,000 predicted binding sites across the genome. Some (`r sum(unlist(lapply(tfbs, length))>1e6)`) have over a million sites, and some (`r sum(unlist(lapply(tfbs, length))<1e4)`) have less than 10 thousand (and maybe these are low-quality PWMs, or factors without a well defined motif).

```{r totalBS}
plot(density(log10(unlist(lapply(tfbs, length)))), bty="l", main="number of TFBS across the mouse geneome", xlab=expression('log'[10]*' binding sites'))
```

However, only a small proportion of these are within regions of open chromatin in somites; around 6% for most TFs, which means there are about 6.5 thousand putative binding sites for each.

```{r total_open}
tfbs.open <- list()
for(tf in names(tfbs)){
  tfbs.open[[tf]] <- tfbs[[tf]][tfbs[[tf]] %over% peakSet]
}
plot(density(log10(unlist(lapply(tfbs.open, length)))), bty="l", main="number of TFBS across the accessible mouse geneome", xlab=expression('log'[10]*' binding sites in open chromatin regions'))
```

However, there are a number of TFs that have a higher proportion of their TFBSs in open regions. Among these are TFs like MBD2, which binds methylated CpGs, which are expected to be enriched in open chromatin due to active promoters (and other TFs binding CpG or GC rich motifs, like SP4 and EGR1). But also factors like KLF6 which is necessary for mesoderm differentiation (KO results in a prolongation of the epiblast-like cell stage and a delay in induction of mesoderm) or E2F1, which has been implicated in regulating muscle formation.

```{r over, warning=FALSE, message=FALSE}
df <- data.frame(tf=names(tfbs.open), open = unlist(lapply(tfbs.open, length))/unlist(lapply(tfbs, length)))

ggplot(df, aes(1, open*100, label=tf)) + geom_violin() + ylab("% of total TFBS in open chromatin") + xlab("") + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) + geom_dotplot(binaxis='y', stackdir='center', dotsize=0.2) + geom_text_repel(data = subset(df, df$open>0.2), nudge_x = 0.05, direction = "y", angle = 0, hjust = 0, segment.size = 0.2)
```

Also important is that nearly a third of all the TFs are not expressed (at significant levels) in somites. And for those that are expressed, there is no relationship between expression level and number of binding sites (or porportion open).

```{r tfs_expr}
## we obtain annotation for the TF PWMs from HOCOMOCO [http://hocomoco10.autosome.ru/final_bundle/MOUSE/mono/HOCOMOCOv10_annotation_MOUSE_mono.tsv]
## we edit this list to match the order of the TFs as read in tfbs. We also remove a few (5) TFs that are not in the diffTF dataset. And we remove ENOA which has been retracted in v11

motif.ann <- read.table(paste0(dir, "ATAC-seq/motifs/HOCOMOCOv10_MOUSE_mono_annotation.tsv"), sep="\t", header = TRUE, stringsAsFactors = FALSE)
stopifnot(identical(names(tfbs.open), motif.ann$diffTF))

## we can now match the TFs to their genes
motif.gene <- motif.ann$Transcription.factor
names(motif.gene) <- motif.ann$diffTF

tf.expr <- rnaseq[match(motif.gene, rnaseq$gene),]$mean
names(tf.expr) <- names(motif.gene)
# sum(is.na(tf.expr)) # 127 (30.1%)

plot(tf.expr, log10(unlist(lapply(tfbs.open, length))), bty="l", xlab=expression('log'[2]*' normalised expression'), ylab=expression('log'[10]*' total binding sites'))
```

For each peak, we can create a matrix with the number of binding sites for each TF that is expressed in the somites. That is `r length(tf.expr[!is.na(tf.expr)])` different TFs across `r length(pepeakSet)` peaks, and thus, the analysis now restricts to expressed TFs and open binding sites.

Peaks have a median of 24 predicted TFBSs, with the IQR between 12 and 42 TFBS.

```{r tfbs_perPeak}
## for each peak, count the number of TFBS
counts.open <- as.data.frame(peakSet)
for(tf in names(tf.expr[!is.na(tf.expr)])){
  counts.open$tf <- 0
  counts.open[unique(queryHits(findOverlaps(peakSet, tfbs[[tf]]))),]$tf <- table(queryHits(findOverlaps(peakSet, tfbs[[tf]])))
  colnames(counts.open)[ncol(counts.open)] <- tf
}
row.names(counts.open) <- paste0(counts.open$seqnames, ":", counts.open$start, "-", counts.open$end)
counts.open$totalTFBS <- rowSums(counts.open[,-c(1:10)])

ggplot(counts.open, aes(1, log10(totalTFBS+1))) + geom_violin() + geom_boxplot(width=0.1) + xlab("") + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) + ylab(expression('log'[10]*' total TFBS + 1'))
nTFs <- table(df$count)
```

`r unname(nTFs['0'])` peaks have no predicted TFBSs and `r sum(nTFs[as.numeric(names(nTFs))>100])` peaks have more than 100 predicted binding sites.

Not surprisingly, peaks that are at promoters have higher numbers of TFBS. This gives us confidence in the computational predictions.

```{r genomicFeatures}
proms <- promoters(gene.ann, upstream = 3000, downstream = 300)
peaks.inPromoters <- peakSet[unique(queryHits(findOverlaps(peakSet, proms)))]
names(peaks.inPromoters) <- paste0(as.data.frame(peaks.inPromoters)[,1],":", as.data.frame(peaks.inPromoters)[,2], "-", as.data.frame(peaks.inPromoters)[,3])

genes.withoutProms <- setdiff(gene.ann, proms)
peaks.inGenes <- peakSet[unique(queryHits(findOverlaps(peakSet, genes.withoutProms)))]
peaks.inGenes <- peaks.inGenes[-unique(queryHits(findOverlaps(peaks.inGenes, peaks.inPromoters, type="equal")))]
names(peaks.inGenes) <- paste0(as.data.frame(peaks.inGenes)[,1],":", as.data.frame(peaks.inGenes)[,2], "-", as.data.frame(peaks.inGenes)[,3])

peaks.intergenic <- peakSet[-unique(queryHits(findOverlaps(peakSet, c(peaks.inGenes, peaks.inPromoters), type = "equal")))]
names(peaks.intergenic) <- paste0(as.data.frame(peaks.intergenic)[,1],":", as.data.frame(peaks.intergenic)[,2], "-", as.data.frame(peaks.intergenic)[,3])

counts.open$prom <- 0
counts.open$gene <- 0
counts.open$inter <- 0

tmp <- as.data.frame(peaks.inPromoters)
counts.open[paste0(tmp$seqnames, ":", tmp$start, "-", tmp$end),]$prom <- 1
tmp <- as.data.frame(peaks.inGenes)
counts.open[paste0(tmp$seqnames, ":", tmp$start, "-", tmp$end),]$gene <- 1
tmp <- as.data.frame(peaks.intergenic)
counts.open[paste0(tmp$seqnames, ":", tmp$start, "-", tmp$end),]$inter <- 1

counts.open$class <- ifelse(counts.open$prom==1, "promoter", ifelse(counts.open$gene == 1, "genic", "intergenic"))

ggplot(counts.open, aes(class, log10(totalTFBS+1))) + geom_violin() + geom_boxplot(width=0.1) + ylab(expression('log'[10]*' total TFBS + 1'))
```

For intergenic peaks, there is no striking differences regarding how far they are from the nearest promoter. There is a slight decrease in the number of TFBS as the distance increases, but this is subtle.

```{r distToNearest}
counts.open$distanceToNearestProm <- 0
counts.open[counts.open$inter==1,]$distanceToNearestProm <- as.data.frame(mcols(distanceToNearest(peaks.intergenic, proms)))[,1]

tmp <- counts.open[counts.open$inter==1, c(1:10,305:310)]
tmp <- tmp[order(tmp$distanceToNearestProm),]
tmp$bin <- c(rep(1:9,each=5190), rep(10, 5229))

plot(sapply(1:10, function(x) summary(tmp[tmp$bin==x,]$totalTFBS))[4,], ylim=c(15,40), pch=16, col="grey", type="b", xlab="increasing distance to nearest promoter", ylab="total number of TFBS", axes=FALSE); box(bty="l"); axis(2); axis(1, at=c(1,10), labels = c("near", "far"))
points(sapply(1:10, function(x) summary(tmp[tmp$bin==x,]$totalTFBS))[3,], pch=16, type="b")
points(sapply(1:10, function(x) summary(tmp[tmp$bin==x,]$totalTFBS))[5,], type="b")
legend("topright", legend = c("75quantile", "mean", "median"), pch=c(1,16,16), col=c("black", "grey","black"), cex=0.8, bty="n")
```

It might be interesting to study intergenic peaks with a large number of TFBSs as potential candidate regulatory enhancers, where we can infer mechanism more easily.

#### HOX factors

We can also study regions with binding sites for a particular TF or family of TFs. Of interest are the Hox genes, which we know are paramount in patterning and defining the identity of the different vertebral segments. However, we don't really understand how they achieve this, or which genes they regulate.

So we take all peaks with a predicted binding site for a `HOX-related factor` family of TFs; this includes all Hox genes plus Cdx1/2 and Evx1/2 (Evx2 is not expressed). Cdx1/2 have around 4 thousand binding sites in peaks, whereas Evx2 and Hox genes have arouns 1 to 2 thousand. In stark contrast, Hoxa1 and Hoxb1 have around 7 thousand sites.

Most of the sites identified as Hoxb1 sites are also called as Hoxa1 sites, since they have very very similar PWMs. This means that these two factors are basically reporting the same features and we are counting them twice. This will also be the case for other Hox sites, where the same site is identified for several Hox factors and thus, we are overestimating the number of TFBS.

In the plot below, if we remove Hoxb1 the numbers pretty much half, indicating that the effect of Hoxa1/b1 is large, given their much higher number of predicted sites.

```{r hox}
tfs <- tf.expr[motif.ann[grep("HOX-related factors", motif.ann$TF.family),]$diffTF]
tfs <- tfs[!is.na(tfs)]

hox.bound <- counts.open[,c("seqnames",names(tfs))]
ggplot(hox.bound, aes(1:nrow(hox.bound), rowSums(hox.bound[,-1]), colour=seqnames)) + geom_point() + scale_color_manual(values = rep(c("black", "grey"), 10)) + theme(legend.position = "none", axis.ticks.x = element_blank(), axis.text.x = element_blank()) + ylab("number of HOX binding sites per peak") + xlab("")
```

Interestingly, if we take all peaks with increasing number of predicted Hox binding sites, we get an increasing proportion of DA peaks across development.

Overall, nearly 20% of the peaks are differential across development. But this goes up to nearly 28% if the peaks have a Hox BS, and this keeps increasing to ~40% if the peaks have more than 6 BSs (though there are only 89 peaks with more than 6 BSs, so the percentage might not be stable).

```{r hox_DA}
nPeaks.hoxBS <- c()
da <- c()
for(i in c(0:8)){
  tmp <- hox.bound[rowSums(hox.bound[,-c(1,7,11)])>=i,]
  tmp <- counts.open[row.names(tmp),]
  nPeaks.hoxBS <- c(nPeaks.hoxBS, nrow(tmp))
  da <- c(da, sum(tmp$DAstages)/nrow(tmp)*100)
}
plot(da, pch=16, ylab="% of peaks DA across stages", xlab="number of HOX binding sites", axes=FALSE)
box(bty="l"); axis(2); axis(1, at=1:8, labels=0:7)
```

**Note:** *for this enalysis I removed both Hoxa1 and Hoxb1 sites. If only one of the two factors is included the result is recapitulated, although the increase is not as linear, but more of a two step process: big jump in the proportion DA when peaks have at least 1 BS, another jump when they have more than 2 sites and increasing the number of sites doesn't increase the % much more (so from 20% to 27% to a stable ~31%). But if both are included, the picture is much less clear.*

We know Hox genes are differentially expressed across development. This tells us that regions targeted by Hox genes, are also more likely to be dynamically opening/closing during development. Whether or not the presence of the Hox proteins has an effect on the chromtin dynamics of these regions is an open question. *They would need to be regulting a chromatin remodeller I suppose.*

So what are these *Hox-bound* regions?

Overall, 20% of open chromatin regions are promoters, an extra 50% are in gene bodies and the remaining 30% are intergenic. The percentages for the subset of peaks with increasing numbers of Hox BSs is similar, but with a slight depletion of promoter and genic peaks, and a slight increase in intergenic peaks. Also subtle but evident, is that the increase in intergenic peaks grows as the peaks are restricted to have more Hox BSs. 

```{r hox_features}
class <- matrix(ncol=4, nrow=9)
for(i in 0:8){
  tmp <- hox.bound[rowSums(hox.bound[,-c(1,7,11)])>=i,]
  tmp <- counts.open[row.names(tmp), c(1:10,305:310)]
  class[i+1,] <- c(nrow(tmp), sum(tmp$prom)/nrow(tmp)*100, 
                   sum(tmp$gene)/nrow(tmp)*100,
                   sum(tmp$inter)/nrow(tmp)*100)
}
cols <- rainbow(n=8)
plot(class[1,-1], type="b", ylim=c(12,60), lwd=2, axes=FALSE, xlab="", ylab="% of peaks")
box(bty="l"); axis(2); axis(1, at=1:3, labels = c("promoter", "genic", "intergenic"))
for(i in 2:nrow(class)){ points(class[i,-1], col=cols[i-1], type="b") }
legend("topleft", legend = c(paste0("all (n=", nPeaks.hoxBS[1], ")"), paste0(">= ", 1:8, " Hox BSs (n=", nPeaks.hoxBS[-1], ")")), col=c("black", cols), pch=16, bty="n", cex=0.75)
```

But there is no difference in the distance of these intergenic peaks to the nearest promoter, compared to all intergenic peaks.

```{r hox_distToNearest}
hox.bound <- counts.open[,c("seqnames",names(tfs), "inter", "distanceToNearestProm")]
hox.bound$nBS <- rowSums(hox.bound[,-c(1,7,11,20,21)])
boxplot(log10(hox.bound[hox.bound$inter==1,]$distanceToNearestProm)~hox.bound[hox.bound$inter==1,]$nBS, xlab="number of Hox BSs", ylab=expression('log'[10]*' distance to nearest promoter'))
```

##### In promoters

From the `r class[2,1]` peaks with at least one Hox binding site, `r class[2,1]*class[2,2]/100` (`r class[2,2]`%) are in promoters. 
```{r}
hox.bound.proms <- peaks.inPromoters[intersect(names(peaks.inPromoters), row.names(hox.bound[hox.bound$nBS>0,]))]
hox.bound.proms <- as.character(proms[unique(subjectHits(findOverlaps(hox.bound.proms, proms)))]$gene) # 5671
```

However, only `r length(intersect(hox.bound.proms, rnaseq$gene))` of these correspond to expresed genes. 

We would expect these genes to be involved in specific functions, if they really are regulated by Hox factors. Gene set enrichment analysis of these `r length(intersect(hox.bound.proms, rnaseq$gene))` genes versus all the `r nrow(rnaseq)` genes expressed in somites returns a number of significant GO terms. 

```{r go_hoxBound, message=FALSE}
universe <- row.names(rnaseq)
sel <- row.names(rnaseq[rnaseq$gene %in% hox.bound.proms,]) # 2992
all <- as.factor(as.numeric(universe %in% sel))
names(all) <- universe

go.hoxBound <- new("topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl")
go.hoxBound.test <- runTest(go.hoxBound, algorithm = "classic", statistic = "Fisher" )
go.hoxBound.res <- GenTable(go.hoxBound, Fisher.classic = go.hoxBound.test, topNodes = length(score(go.hoxBound.test)), numChar=60)
go.hoxBound.res$Fisher.classic.adj <- p.adjust(go.hoxBound.res$Fisher.classic, "fdr")
```
There are `r sum(go.hoxBound.res$Fisher.classic < 0.01)` terms with an enrichment p-value lower than 0.01. However, this drops to `r sum(go.hoxBound.res$Fisher.classic.adj < 0.05)` if we consider an FDR of 5%. Nonetheless, there are some interesting terms within the top 100:

```{r terms_hoxBound}
go.hoxBound.res[c(1:6,8:9,13,19,28,48,53,62,79,90,94,96),-c(1,5)]
```

This suggests that the promoters we have identified as possibly bound by Hox factors make biological sense; again, giving us confidence on the computational predictions of TFBSs.

Taking this further, we have identified a number of genes with their expression levels significantly correlated to that of Hox factors. We would expect that those that are correlated because they are regulated by Hox factors would have Hox BSs in their promoters. 

```{r hox_corrGenes}
hox.corr.genes <- read.table(paste0(dir, "RNA-seq/results/HoxCorrelatedGenes.tab"), stringsAsFactors = FALSE)
hox.corr.genes$HOXbound <- ifelse(hox.corr.genes$gene %in% intersect(hox.corr.genes$gene, hox.bound.proms), 1, 0)
```

From `r length(unique(hox.corr.genes$gene2))` genes correlated with Hox expression, `r length(unique(hox.corr.genes[hox.corr.genes$HOXbound==1,]$gene2))` (`r round(length(unique(hox.corr.genes[hox.corr.genes$HOXbound==1,]$gene2))/length(unique(hox.corr.genes$gene2))*100,2)`%) have HOX binding sites in their promoters. The remaining could still be regulated by Hox factors via enhancers, but it is likely that a large fraction of these are not directly regulated by HOX genes, and instead are correlated becuase they are following the developmental progression of somites.

The correlated genes with HOX binding sites in their promoters are biased to positive correlations, much more than genes without HOX binding sites.

```{r positiveCorr}
tmp <- data.frame(noHOX = as.vector(round(table(hox.corr.genes[hox.corr.genes$HOXbound==0,]$rho>0)/sum(hox.corr.genes$HOXbound==0)*100,2)), HOX = as.vector(round(table(hox.corr.genes[hox.corr.genes$HOXbound==1,]$rho>0)/sum(hox.corr.genes$HOXbound==1)*100,2)))
row.names(tmp) <- c("negative correlation", "positive correlation")
t(tmp)
```

Now we can test if these couple hundred genes are enriched for particular functions. 

```{r go_hoxCorr, message=FALSE}
sel <- hox.corr.genes[hox.corr.genes$HOXbound==1,]$gene2
all <- as.factor(as.numeric(universe %in% sel))
names(all) <- universe

go.hoxCorr <- new( "topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl" )
go.hoxCorr.test <- runTest(go.hoxCorr, algorithm = "classic", statistic = "Fisher" )
go.hoxCorr.res <- GenTable(go.hoxCorr, Fisher.classic = go.hoxCorr.test, topNodes = length(score(go.hoxCorr.test)), numChar=60)
go.hoxCorr.res$Fisher.classic.adj <- p.adjust(go.hoxCorr.res$Fisher.classic, "fdr")
```

In this case, `r sum(go.hoxCorr.res$Fisher.classic < 0.01)` terms have an enrichment p-value lower than 0.01 and `r sum(go.hoxCorr.res$Fisher.classic.adj < 0.05)` of these are significant under an FDR of 5%. Again, many interesting terms within the top 100:

```{r terms_hoxCorr}
go.hoxCorr.res[c(6:8,16,24,26:28,36,40,47,71,73:76,80,95),-c(1,5)]
```

Terms related to muscle and bone processes are much more significant compared to the test with all genes with Hox sites in their pormoters. This suggests that by taking only correlated genes, we have identified a subset of genes involved in more specific functions.

As a control, we can test for enrichments in the set of correlated genes without Hox binding sites.

```{r go_hoxCorr_unbound, message=FALSE}
sel <- hox.corr.genes[hox.corr.genes$HOXbound==0,]$gene2
all <- as.factor(as.numeric(universe %in% sel))
names(all) <- universe

go.ctrl <- new( "topGOdata", ontology="BP", allGenes = all, nodeSize=5, annot=annFUN.org, mapping="org.Mm.eg.db", ID = "ensembl" )
go.ctrl.test <- runTest(go.ctrl, algorithm = "classic", statistic = "Fisher" )
go.ctrl.res <- GenTable(go.ctrl, Fisher.classic = go.ctrl.test, topNodes = length(score(go.ctrl.test)), numChar=70)
go.ctrl.res$Fisher.classic.adj <- p.adjust(go.ctrl.res$Fisher.classic, "fdr")
```

In this case, `r sum(go.ctrl.res$Fisher.classic < 0.01)` terms have an enrichment p-value lower than 0.01 but `r sum(go.ctrl.res$Fisher.classic.adj < 0.05)` of these are significant under an FDR of 5%. From the top 50 terms, enrichment is for more general terms, related to metabolism and general cellular functions.

```{r terms_hoxCorr}
go.ctrl.res[c(2,6,10,22,25,28,31,35,36,44),-c(1,5)]
```

Thus, by selecting only correlated genes that have Hox binding in their promoters we have certainly captured a set of biological interest. It is related to the functions of all genes with Hox binding in their promters, but we have the added information of correlated expression, suggesting more plausible regulation.


```{r}
selected <- c(6:8,16,24,26:28,36,40,47,71,73:76,80,95)
go.hox <- matrix(ncol=18,nrow = length(selected))
j=1
for(i in selected){
  genes <- proms[proms$ID %in% intersect(genesInTerm(go.hoxCorr, go.hoxCorr.res[i,1])[[1]], hox.corr.genes[hox.corr.genes$HOXbound==1,]$gene2)]
  tmp <- hox.bound[names(peaks.inPromoters[unique(queryHits(findOverlaps(peaks.inPromoters, genes)))]),]
  tmp <- tmp[tmp$nBS>0,]
  go.hox[j,] <- colSums(tmp[,2:19])
  j <- j+1
}
colnames(go.hox) <- colnames(tmp)[2:19]
row.names(go.hox) <- go.hoxCorr.res[selected,2]
go.hox

heatmap.2(go.hox, trace = "none")

```



##### In genes


##### Intergenic






#### HOX co-factors

```{r}

tf.expr[substr(motif.ann[grep("homeo domain", motif.ann$TF.family),]$Model, 1, nchar(motif.ann[grep("homeo domain", motif.ann$TF.family),]$Model)-14)]
```




```{r info}
sessionInfo()
```

