---
title: "Transcription factor binding sites"
date: "15 April 2019"
output:
  html_document:
    keep_md: true
    fig_width: 5
    fig_height: 5
    fig_caption: yes
    code_folding: hide
    toc: true
    toc_float: 
      collapsed: false
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(GenomicRanges)
library(csaw)
library(ChIPpeakAnno)
library(TxDb.Mmusculus.UCSC.mm10.knownGene)
library(org.Mm.eg.db)
library(ggplot2)
library(ggrepel)
library(gplots)
library(RColorBrewer)
dir <- "/user01/group_folders/Personal/Ximena/SOMITES/somitogenesis2019/"

#### FUNCTIONS
multiplot <- function(..., plotlist=NULL, file, cols=1, layout=NULL) {
  library(grid)
  # Make a list from the ... arguments and plotlist
  plots <- c(list(...), plotlist)
  numPlots = length(plots)
  
  # If layout is NULL, then use 'cols' to determine layout
  if (is.null(layout)) {
    # Make the panel
    # ncol: Number of columns of plots
    # nrow: Number of rows needed, calculated from # of cols
    layout <- matrix(seq(1, cols * ceiling(numPlots/cols)),
                     ncol = cols, nrow = ceiling(numPlots/cols), byrow = TRUE)
  }
  
  if (numPlots==1) {
    print(plots[[1]])
    
  } else {
    # Set up the page
    grid.newpage()
    pushViewport(viewport(layout = grid.layout(nrow(layout), ncol(layout))))
    
    # Make each plot, in the correct location
    for (i in 1:numPlots) {
      # Get the i,j matrix positions of the regions that contain this subplot
      matchidx <- as.data.frame(which(layout == i, arr.ind = TRUE))
      
      print(plots[[i]], vp = viewport(layout.pos.row = matchidx$row,
                                      layout.pos.col = matchidx$col))
    }
  }
}
```

```{r meatadata}
## metadata
meta <- read.table(paste0(dir, "ATAC-seq/data/metadata_ATACseq.tsv"), stringsAsFactors = FALSE, header = TRUE)
meta <- meta[meta$QCpass==1,]

## peaks
peakSet <- readRDS(file=paste0(dir, "ATAC-seq/results/04.5_peakSet_csawMerged_geneAnn.Rds"))

## gene expression
rnaseq <- read.table(paste0(dir, "RNA-seq/data/normalisedCounts_batchCorrected.tab"))
rnaseq$mean <- apply(rnaseq[,-1], 1, mean)

gene.ann <- read.table(paste0(dir, "RNA-seq/data/Mus_musculus.GRCm38.93.ann"))
gene.ann <- GRanges(paste0("chr", gene.ann$V3), IRanges(gene.ann$V4, gene.ann$V5), strand = gene.ann$V6, gene=gene.ann$V2, ID=gene.ann$V1)
```


```{r readTFmatches_mm10}
## HOCOMOCO v10 PWMs used to scan the mm10 genome with PWMscan, default parameters.
## Taken from diffTF [https://www.embl.de/download/zaugg/diffTF/TFBS/TFBS_mm10_PWMScan_HOCOMOCOv10.tar.gz]
tfbs <- list()
files <- list.files(paste0(dir, "ATAC-seq/motifs/PWMScan_HOCOMOCOv10_mm10/"))
for(file in files){
  tf <- substr(file, 1, nchar(file)-9)
  tfbs[[tf]] <- read.table(paste0(dir, "ATAC-seq/motifs/PWMScan_HOCOMOCOv10_mm10/", file))
  tfbs[[tf]] <- GRanges(tfbs[[tf]]$V1, IRanges(tfbs[[tf]]$V2, tfbs[[tf]]$V3), strand=tfbs[[tf]]$V6, motif=tfbs[[tf]]$V4, score=tfbs[[tf]]$V5)
}
saveRDS(tfbs, paste0(dir, "ATAC-seq/motifs/PWMScan_HOCOMOCOv10_mm10.Rds"))
```

We will use the PWMs from HOCOMOCO, which is based on ChIP-seq data and should contain high-quality motifs. Each PWM is used to scan the mouse refence genome and predict TF binding sites. We read the coordinates of such putative TFBSs for all `r length(tfbs)` TFs available.

Most TFs have around 100,000 predicted bindingi sites across the genome. Some (`r sum(unlist(lapply(tfbs, length))>1e6)`) have over a million sites, and some (`r sum(unlist(lapply(tfbs, length))<1e4)`) have less than 10 thousand (and maybe these are low-quality PWMs, or factors without a well defined motif).

```{r totalBS}
plot(density(log10(unlist(lapply(tfbs, length)))), bty="l", main="number of TFBS across the mouse geneome", xlab=expression('log'[10]*' binding sites'))
```

However, only a small prportion of these are within regions of open chromatin in somites; around 6% for most TFs, which means there are about 6.5 thousand putative binding sites for each.

```{r total_open}
tfbs.open <- list()
for(tf in names(tfbs)){
  tfbs.open[[tf]] <- tfbs[[tf]][tfbs[[tf]] %over% peakSet]
}
plot(density(log10(unlist(lapply(tfbs.open, length)))), bty="l", main="number of TFBS across the accessible mouse geneome", xlab=expression('log'[10]*' binding sites in open chromatin regions'))
```

However, there are a number of TFs that have a higher proportion of their TFBSs in open regions. Among these are TFs like MBD2, which binds methylated CpGs, which are expected to be enriched in open chromatin due to active promoters (and other TFs binding CpG or GC rich motifs, like SP4 and EGR1). But also factors like KLF6 which is necessary for mesoderm differentiation (KO results in a prolongation of the epiblast-like cell stage and a delay in induction of mesoderm) or E2F1, which has been implicated in regulatin muscle formation.

```{r over, warning=FALSE, message=FALSE}
df <- data.frame(tf=names(tfbs.open), open = unlist(lapply(tfbs.open, length))/unlist(lapply(tfbs, length)))

ggplot(df, aes(1, open*100, label=tf)) + geom_violin() + ylab("% of total TFBS in open chromatin") + xlab("") + theme(axis.text.x = element_blank(), axis.ticks.x = element_blank()) + geom_dotplot(binaxis='y', stackdir='center', dotsize=0.2) + geom_text_repel(data = subset(df, df$open>0.2), nudge_x = 0.05, direction = "y", angle = 0, hjust = 0, segment.size = 0.2)
```

Also important is that nearly a third of all the TFs are not expressed (at significant levels) in somites. And for those that are expressed, there is no relationship between expression level and number of binding sites (or porportion open).

```{r tfs_expr}
motif.ann <- read.table(paste0(dir, "ATAC-seq/motifs/HOCOMOCOv10_MOUSE_mono_annotation.tsv"), sep="\t", header = TRUE, stringsAsFactors = FALSE)
## there are a few TFs missing in the diffTF data, so need to remove from the annotation file
motif.ann <- motif.ann[-c(23,30,256,258,395),]

motif.gene <- motif.ann$Transcription.factor
names(motif.gene) <- names(tfbs)

motif.gene
tf.expr <- rnaseq[match(motif.gene, rnaseq$gene),]$mean
names(tf.expr) <- names(tfbs.open)
# sum(is.na(tf.expr)) # 127 (30.1%)

plot(tf.expr, log10(unlist(lapply(tfbs.open, length))), bty="l", xlab=expression('log'[2]*' normalised expression'), ylab=expression('log'[10]*' total binding sites'))
```



```{r}
## tile the genome; 1Mb regions
tiles <- tileGenome(seqinfo(BSgenome.Mmusculus.UCSC.mm10), tilewidth = 1e6, cut.last.tile.in.chrom = TRUE)

## for each TF, count the number of BS per tile
counts <- data.frame(tile = 1:length(tiles), chr = seqnames(tiles))

plots <- list()
for(tf in names(tfbs)){
  counts$tf <- 0
  counts[unique(queryHits(findOverlaps(tiles, tfbs[[tf]]))),]$tf <- table(queryHits(findOverlaps(tiles, tfbs[[tf]])))
  colnames(counts)[ncol(counts)] <- tf
  plots[[tf]] <- ggplot(counts, aes(tile, counts[,colnames(counts)[ncol(counts)]], color=chr)) + geom_point() + theme(legend.position = "none") + scale_colour_manual(values = rep(c("black", "grey"),33)) + ylab("total TFBS per Mb") + xlab("") + theme(axis.ticks.x = element_blank(), axis.text.x = element_blank()) + ggtitle(colnames(counts)[ncol(counts)])
}
head(counts)
```












```{r info}
sessionInfo()
```

